extension-bolt: simple taproot channels (feature 80/81) (#995)
lightningnetwork/lightning-rfc: bolt-simple-taproot.md

designs.

## Preliminaries

In this section we lay out some preliminary concepts, protocol flows, and
notation that'll be used later in the core channel specification.

### Pay-To-Taproot-Outputs

The Taproot soft fork package (BIPs [340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki), [341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki), and [342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki)) introduced a new [Segwit](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki)
witness version: v1. This new witness version utilize a new standardized public
key script template:
```
OP_1 <taproot_output_key>
```

The `taproot_output_key` is a 32-byte x-only `secp256k1` public key, with all
digital signatures based off of the schnorr signature scheme described in BIP
340.

A `taproot_output_key` commits to an internal key, and optional script root via
the following mapping:
```
taproot_output_key = taproot_internal_key + tagged_hash("TapTweak", taproot_internal_key || script_root)*G
```

It's important to note that while `taproot_output_key` is serialized as a

> 32-byte public key, in order to properly spend the script path, the parity

Reviewer: can say the parity bit of the output pubkey `Q`

(even or odd) of the public key must be remembered.

The `taproot_internal_key` is also a BIP 340 public key, ideally derived anew
for each output. The `tagged_hash` scheme is described in BIP 340, we briefly
define the function as:
```
tagged_hash(tag, msg) = SHA256(SHA256(tag) || SHA256(tag) || msg)
```

#### Tapscript Tree Semantics

The `script_root` is the root of a Tapscript tree as defined in BIP 341. A tree
is composed of `tap_leaves` and `tap_branches`.

A `tap_leaf` is a two tuple of (`leaf_version`, `leaf_script`). A `tap_leaf` is serialized as:
```
leaf_version || compact_size_of(leaf_script) || leaf_script
```
where `compact_size_of` is the variable length integer used in the Bitcoin p2p
protocol.

The digest of a `tap_leaf` is computed as:
```

> tagged_hash("TapLeaf", leaf_encoding)

Reviewer: here `leaf_encoding` not defined, I think `v || compact_size(size of s) || s` or `leaf_version || compact_size_of(leaf_script) || leaf_script` is meant

```
<key_path_sig>
```

If an output commits to a script root (or uses BIP 86), a private key will need
to be tweaked in order to generate a valid signature, and negated if the
resulting output key has an odd y coordinate (see BIP 341 for details).

#### Taproot Script Path Spends

A script path spend is executed when a tapscript leaf is revealed. A script
path spends has three components: a control block, the script leaf being
revealed, and the valid witness.

A control block has the following serialization:
```
(output_key_y_parity | leaf_version) || internal_key || inclusion_proof
```

The first segment uses a spare bit of the `leaf_version` to encode the parity
of the output key, which is checked during control block verification. The
`internal_key` is just that. The `inclusion_proof` is the series of sibling
hashes in the path from the revealed leaf all the way up to the root of the
tree. In practice if one hashes the revealed leaf, and each 32-byte hash of the
inclusion proof together, they'll reach the script root if the proof was valid.

If only a single leaf is committed to in a tree, then the `inclusion_proof`
will be absent.

> The final witness to spend a script path output is:

Reviewer: can say "final witness stack" in bip141 parlance

    aggregated public key.

  3. Next, each party generates a _partial_ signature and exchanges it with
    every other party.

  4. Finally, once all partial signatures are known, they can be combined into a
    valid schnorr signture, which can be veirfied using BIP 340.

Steps 1&2 may be performed out of order, or concurrently. In our case only two
parties exist, so as soon as one party knows both partial signatures, they can
be combined into a final signature.

#### Key Aggregation

We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)
for key aggregation. In order to avoid sending extra ordering information, we
always assume that both keys are sorted first with the `KeySort` algorithm
before aggregating (`KeyAgg(KeySort(p1, p2))`).

#### Nonce Generation

A `musig2` secret nonce is the concatenation of two random, 32-byte integers.

A `musig2` public nonce is technically the concatenation of two public keys, each representing the EC-point
corresponding to its secret integer, thus resulting in a 66-byte value:

```
point_1 || point_2
```

For nonce generation, we refer to the `NonceGen` algorithm defined in the

> `musig2` BIP. We only specify the set of required arguments, though a signer can also opt to specify the additional

Reviewer: Should we put "Nonce generation" recommendations somewhere, e.g committing to the `funding_outpoint || per_commitment_point` in `sec256k1_musig_nonce_gen`'s `extra_input32` ? Like per-commitment point should be unique for each channel state which is nice (unless you have `channel_reestablish`).

Author: Yep, I've tacked this on in the latest version.

counterparty to create a valid partial signature.

The nonce for the remote commitment _could_ be sent a priori, too, but for the
purposes of simplicity and disambiguation, MUST be sent alongside the partial
signature. To that end, we will create a new data type that will be reused in
multiple TLV extensions described further down in this document.

The exception to this is the co-op close flow: as there's only a single message
to sign, we only require a single pair of nonces. These nonces can also be sent
right as the co-op flow begins.

#### Signing

Once public nonces have been exchanged, the `NonceAgg` algorithm is used to
combine the public nonces into the aggregate nonce which will be a part of the
final signature.

The `Sign` method is used to generate a valid partial signature, with the
`PartialSigVerify` algorithm used to verify each signature.

Once all partial signatures are known, the `PartialSigAgg` algorithm is used to
aggregate the partial signature into a final, valid [BIP
340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki) Schnorr
signature.

##### Partial Signature Encoding

> A standard `musig2` partial signature is simply the encoding of the `s` value,

Author: I sort of leave it up to the BIP to define this value.

which is a 32-byte-element modulo the order of the group.

Throughout this specification, we will be using a custom type,
`PartialSignatureWithNonce`, comprised of the aforementioned `s` value, along
with the 66-byte-encoding of the public nonce used to sign remote commitments:

```
s || point_1 || point_2
```

## Design Overview

With the preliminaries out of the way, we provide a brief overview of the
Simple Taproot Channel design.

The multisig output becomes a single P2TR key, with `musig2` key aggregation
used to combine two keys into one.

For commitment transactions, we inherit the anchor output semantics, meaning
there are two outputs used for CPFP, with all other scripts inheriting a `1
CSV` restriction.

The local output of the commitment transaction uses the revocation key as the
taproot internal key, and commits to a single script path of the delay script.

> The remote output of the commitment transaction uses the `combined_funding_key`

Reviewer: If the `combined_funding_key` is never used iiuc the new flow, I think it would be better to make it a NUMS point as described in bip340 or bip341 iirc. Otherwise, I would be worried about future uprades where APO is introduced and someone can re-bind a `combined_funding_key` spent on this output, double-spending the remote output.

Author: Yep, latest version uses a NUMs point now, we had one, then went back to it.

as the top-level internal key, and then commits to a normal remote script.

Anchor outputs use the respective two parties' funding keys as the top-level
internal key committing to the script `16 CSV`. If a local or remote output is
missing from the commitment, its corresponding anchor must not be present,

> either.

Reviewer: If they are HTLCs, even if no remote output, an anchor output must be present for the remote party according to today BOLT3: https://github.com/lightning/bolts/blob/master/03-transactions.md#to_local_anchor-and-to_remote_anchor-output-option_anchors (you have pending funds at stake!)

Author: Good point, clarified.

All HTLC scripts use the revocation key as the top-level internal key. This
allows the revocation to be executed without additional on-chain bytes, and
also reduces the amount of state that nodes need to keep in order to properly
handle channel breaches.

## Specification

### Feature Bits

Inheriting the structure put forth in BOLT 9, we define a new feature bit to be
placed in the `init` message, `node_announcement` and also
`channel_announcement`. We opt to also place the bit in the
`channel_announcement` so taproot channels can be identified on the routing
level, which will eventually matter once PTLCs are fully rolled out.

| Bits  | Name             | Description                           | Context | Dependencies                             | Link                 |
|-------|------------------|---------------------------------------|---------|------------------------------------------|----------------------|

> | 30/31 | `option_taproot` | Node supports simple taproot channels | IN      | `option_channel_type` + `option_anchors` | TODO(roasbeef): link |

Reviewer: can say `option_anchors_zero_fee_htlc_tx` as the other one is unsafe (even if more verbose though better than an ill-implemented channel negotiation logic setting legacy option_anchors for simple taproot channels :) )

Author: IIRC, we made a change to just always refer to the final anchor as simpluy `option_anchors`.

> | 30/31 | `option_taproot` | Node supports simple taproot channels | IN      | `option_channel_type` + `option_anchors` | TODO(roasbeef): link |

Reviewer 2: i dont think it makes sense to have a bit in the `channel_announcement`. New gossip messages will be needed to advertise taproot channels & so the feature bit can just be implied from the fact that it was advertised using the new message

> | 30/31 | `option_taproot` | Node supports simple taproot channels | IN      | `option_channel_type` + `option_anchors` | TODO(roasbeef): link |

Author: Have it just has `IN` here since now we have Gossip 1.75 to handle advertising these channels.

#### `open_channel` Extensions

1. `tlv_stream`: `open_channel_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:
      * [`66*byte`:`public_nonce`]

##### Requirements

The sending node:

  - MUST specify the `next_local_nonce` field.
  - MUST use the `NonceGen` algorithm defined in `bip-musig2` to generate
    `next_local_nonce` to ensure it generates nonces in a safe manner.

The receiving node MUST fail the channel if:

  - the message doesn't include a `next_local_nonce` value.

  - the specified public nonce cannot be parsed as two compressed secp256k1
    points

##### Rationale

For the initial Taproot channel type (`option_taproot`), musig2 nonces
must be exchanged in the first round trip (open->, <-accept) to ensure that the
initiator is able to generate a valid musig2 partial signature at the next
step once the transaction to be signed is fully specified.

The `next_local_nonce` field will be used to verify incoming commitment
signatures. Each incoming signature will carry the newly generated nonce used
to sign the commitment transaction. At force close broadcast time, the
verification nonce is then used to sign the local party's commitment

> transaction.

Author: >  is sent for the remote party to use in signing our local commitment transaction.
> It is NOT used for verifying incoming commitment signatures, in contrast to what is described here.

These are duals of the same operation: remote party sends a sig using the nonce, I verify the sig using that same nonce.

#### `accept_channel` Extensions

1. `tlv_stream`: `accept_channel_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:
      * [`66*byte`:`public_nonce`]

##### Requirements

The sender:

  - MUST set `next_local_nonce` to the `musig2` public nonce used to sign local
    commitments as specified by the `NonceGen` algorithm of `bip-musig2`.

The recipient:

  - MUST reject the channel if `next_local_nonce` is absent.

  - the specified public nonce cannot be parsed as two compressed secp256k1
    points

#### `funding_created` Extensions

1. `tlv_stream`: `funding_created_tlvs`
2. types:
   1. type: 2 (`partial_signature_with_nonce`)
   2. data:

>       * [`98*byte`: `partial_signature || public_nonce`]

Author: They're to distinct types: one with the nonce and one without. Having to types makes the context clearer: it isn't possible to specify an incomplete message (has the sig, not the nonce, but should also have the nonce).

        * [`66*byte`:`nonces`]

Before a signature can be generated for a co-op close transaction, but sides
must exchange a fresh pair of `musig2_pubnonce` values. We package this with
the `shutdown` message so that both sides can send a `closing_signed` message
as soon as a `shutdown` message is sent and received.

Only a single nonce is needed as there's only a single message to sign: the
shared co-op close transaction.

##### Requirements

A sending node:

 - MUST set the `shutdown_nonce` to a valid `musig2` public nonce.

A receiving node:

 - MUST verify that the `shutdown_nonce` value is a valid `musig2` public
   nonce.

#### `closing_signed` Extensions

We add a new TLV to the `closing_signed`
message's existing `tlv_stream`:

1. `tlv_stream`: `closing_signed_tlvs`
2. types:
   1. type: 6 (`partial_signature`)
   2. data:
      * [`32*byte`: `partial_signature`

Both sides **MUST** provide this new TLV field.

Once all partial signatures are known, the `PartialSigAgg` algorithm is used to
aggregate the partial signature into a final, valid [BIP
340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki) Schnorr
signature.

#### Requirements

The sender:

>   - MUST set the original, non-TLV `signature` field to a 0-byte-array of length

Author: > why cant we drop the unused fields since these messages wont ever need to be parsed by nodes not compatible with this spec

Less changes, everyone is able to re-use all their parsing, and just read the new TLV fields. 

>  Further, why not change all messages in the funding flow after open_channel & accept_channel to be purely tlv?

Less changes, the messages are nearly entirely the same. All the existing fields have their same meaning, we just now add the nonce + sig fields.

  - MUST compute the aggregate nonce from:

    - the `next_local_nonce` field the recipient previously sent in the latest
      of `channel_ready`, `channel_reestablish`, or `revoke_and_ack`

    - the `public_nonce` included as part of the `partial_signature_with_nonce`
      field

  - MUST verify the `partial_signature_with_nonce` field using the
    `PartialSigVerifyInternal` algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

  - MUST fail the channel if _any_ of the received HTLC signatures is invalid
    according to BIP 342.

#### `revoke_and_ack` Extensions

A new TLV stream is added to the `revoke_and_ack` message:

1. `tlv_stream`: `revoke_and_ack_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:
      * [`66*byte`: `public_nonce`]

Similar to sending the `next_per_commitment_point`, we also send the _next_
`musig2` nonce after we revoke a state. Sending this nonce allows the remote
party to propose another state transition as soon as the message is received.

##### Requirements

The sender:

- MUST generate a unique nonce to send in the `next_local_nonce` field.

The recipient:

> - MUST fail the channel if `next_local_nonce` is absent.
> 

Reviewer 3: ```suggestion
- MUST fail the channel if `next_local_nonce` is absent or cannot be parsed as
  two compressed secp256k1 points.
```

- If the local nonce generation is non-deterministic and the recipient co-signs
  commitments only upon pending broadcast:

  - MUST **securely** store the local nonce.

### Message Retransmission

#### `channel_reestablish` Extensions

We add a new TLV field to the `channel_reestablish` message:

1. `tlv_stream`: `channel_reestablish_tlvs`
2. types:
1. type: 4 (`next_local_nonce`)
2. data:
   * [`66*byte`: `public_nonce`]

Similar to the `next_per_commitment_point`, by sending the `next_local_nonce`
value in this message, we ensure that the remote party has our public nonce
which is required to generate a new commitment signature.

##### Requirements

The sender:

- MUST set `next_local_nonce` to a fresh, unique `musig2` nonce as specified by
  `bip-musig2`

The recipient:

> - MUST fail the channel if `next_local_nonce` is absent.

Reviewer 3: ```suggestion
- MUST fail the channel if `next_local_nonce` is absent or cannot be parsed as
  two compressed secp256k1 points
```

aggregated public key.

The resulting funding output script takes the form:
`OP_1 funding_key`

where:

  * `funding_key = combined_funding_key + tagged_hash("TapTweak", combined_funding_key)*G`
  * `combined_funding_key = musig2.KeyAgg(musig2.KeySort(pubkey1, pubkey2))`

The funding key is derived via the recommendation in BIP 341 that states: "If
the spending conditions do not require a script path, the output key should
commit to an unspendable script path instead of having no script path. This can
be achieved by computing the output key point as `Q = P +
int(hashTapTweak(bytes(P)))G`". We refer to BIP 86 for the specifics of this
derivation.

By using the musig2 sorting routine, we inherit the lexicographical key
ordering from the base segwit channels. Throughout the lifetime of a channel
any keys aggregated via musig2 also are assumed to use the `KeySort` routine

### Commitment Transactions

#### To Local Outputs

For the simple taproot commitment type, we use the same flow of revocation or
delay, but instead left the revocation case into the taproot key-spend case.

The new output has the following form:

  * `OP_1 to_local_output_key`
  * where:

>     * `to_local_output_key = revocationpubkey + tagged_hash("TapTweak", revocationpubkey || to_delay_script_root)`

Reviewer 2: nit: missing `* G` after the tweak

    * `to_delay_script_root = tapscript_root([to_delay_script])`
    * `to_delay_script` is the delay script:
        ```
        <local_delayedpubkey> OP_CHECKSIG

>         <to_self_delay> OP_CHECKSEQUENCEVERIFY OP_DROP

Reviewer 4: [the shortest form possible](https://github.com/Roasbeef/lightning-rfc/pull/1#discussion_r886414660) has 3 different scripts (0=no checksequenceverify, 1=use the 1 from checksig like in the to_remote script, else the script that was before my pr in the bolt proposal), but this would get too complex.
using this one is possible and saves a byte, but keep in mind the counterparty decides your csv_delay (could be 0), and an unspendable script would be a too high risk (accepting a 0 local csv due to a bug or an oversight of the ln impl or some other tool)

>         <to_self_delay> OP_CHECKSEQUENCEVERIFY OP_DROP

Author: That might be shorter from the PoV of the individual scripts, but the added scripts increases the size of the control block (you can make one of them shorter though).

>         <to_self_delay> OP_CHECKSEQUENCEVERIFY OP_DROP

Reviewer 4: i dont mean to include every script in the taptree, the node chooses (and the remote node enforces) that the right script is used
if csv_delay is 0: `<pubkey> OP_CHECKSIG`
if csv_delay is 1: `<pubkey> OP_CHECKSIG OP_CHECKSEQUENCEVERIFY`
else: `<pubkey> OP_CHECKSIGVERIFY <csv_delay> OP_CHECKSEQUENCEVERIFY`

>         <to_self_delay> OP_CHECKSEQUENCEVERIFY OP_DROP

Reviewer 5: without being overly cute, switching to the 0/non-zero cases above 

e.g.
```
if csv_delay is 0: <pubkey> OP_CHECKSIG
else: <pubkey> OP_CHECKSIGVERIFY <csv_delay> OP_CHECKSEQUENCEVERIFY
```
makes things miniscript compatible and is trivial enough?

along side `<revocationpubkey>` to derive the private key needed to sweep the
top-level key spend path. A valid witness is then just:
```
<revocation_sig>
```

In the case of a routine force close, the script path must be revealed so the
broadcaster can sweep their funds after a delay. The control block to spend is
only `33` bytes, as it just includes the internal key (along with the y-parity
bit and leaf version):
```
delay_control_block = (output_key_y_parity | 0xc0) || revocationpubkey
```

A valid witness is then:
```
<local_delayedsig> <to_delay_script> <delay_control_block>
```

As with base channels, the `nSequence` field must be set to `to_self_delay`.

#### To Remote Outputs

As we inherit the anchor output semantics we want to ensure that the remote
party can unilaterally sweep their funds after the 1 block CSV delay. In order
to achieve this property, we'll re-use the `_funding_key` here: its in
the best interest of the other party to enforce these semantics (mempool
pinning mitigation), so the remote party will be forced to always take the
script reveeal path.

The to remote output has the following form:

  * `OP_1 to_remote_output_key`
  * where:

>     * `to_remote_output_key = combined_funding_key + tagged_hash("TapTweak", combined_funding_key || to_remote_script_root)`

Reviewer 2: `* G` also missing here

    * `to_remote_script_root = tapscript_root([to_remote_script])`
    * `to_remote_script` is the remote script:
        ```
        <remotepubkey> OP_CHECKSIG

>         OP_CHECKSEQUENCEVERIFY

Reviewer 4: for legibility reasons a comment can just be added here explaining why this works

        ```

This output can be swept by the remote party with the following witness:
```

> <remote_sig> <to_remote_script> <to_remote_control_block>

Reviewer: the `nSequence` field must be set to 1 of the spend input

    * `anchor_internal_key = remotepubkey/local_delayedpubkey`
    * `anchor_output_key = anchor_internal_key + tagged_hash("TapTweak", anchor_internal_key || anchor_script_root)`
    * `anchor_script_root = tapscript_root([anchor_script])`
    * `anchor_script`:
        ```
        OP_16 OP_CHECKSEQUENCEVERIFY
        ```

This output can be swept by anyone after 16 blocks with the following witness:
```
<anchor_script> <anchor_control_block>
```

where `anchor_control_block` is:
```
(output_key_y_parity | 0xc0) || anchor_internal_key
```

`nSequence` needs to be set to 16.

### HTLC Scripts & Transactions

#### Offered HTLCs

We retain the same second-level HTLC mappings as base channels. The main
modifications are using the revocation public key as the internal key, and
splitting the timeout and success paths into individual script leaves.

An offered HTLC has the following form:

  * `OP_1 offered_htlc_key`
  * where:
    * `offered_htlc_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
    * `htlc_script_root = tapscript_root([htlc_timeout, htlc_success])`
    * `htlc_timeout`:
        ```
        <local_htlcpubkey> OP_CHECKSIGVERIFY

>         <remote_htlcpubkey> OP_CHECKSIG

Reviewer 6: Thanks to @ellemouton great [article](https://ellemouton.com/posts/taproot-chan-txs/), I realized that this script is missing a  OP_CHECKSEQUENCEVERIFY

        ```
    * `htlc_success`:
        ```

>         OP_SIZE 32 OP_EQUALVERIFY OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY

Author: No, this is what prevent someone from creating a 50 KB pre-image that may has issues being propagated or mined on chain for proper reveal.

>         OP_SIZE 32 OP_EQUALVERIFY OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY

Reviewer 4: this would also allow weird situations, and maybe partial settlement attacks because the lightning node only accepts 32 byte preimages, but on chain (when force closing) this 32 byte requirement is no longer there. routing nodes would not be sure anymore that preimages they extract from an on chain spend can be used down the route

>         OP_SIZE 32 OP_EQUALVERIFY OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY

Reviewer 5: I think making the preimage larger than 80 bytes would make it non-standard today(see MAX_STANDARD_TAPSCRIPT_STACK_ITEM_SIZE in Core repo). This would allow theft at fairly low cost with cooperation of a miner.

        <remote_htlcpubkey> OP_CHECKSIG

>         OP_CHECKSEQUENCEVERIFY

Reviewer 5: my 2 cents only but don't think saving 1 WU per onchain success tx is worth it. miniscript does the expected variant:
```
and(hash160(H),and(pk(A),after(1)))
```

```
OP_SIZE <20> OP_EQUALVERIFY OP_HASH160 <h> OP_EQUALVERIFY <A> OP_CHECKSIGVERIFY
1 OP_CHECKLOCKTIMEVERIFY
```

Reviewer 4: was discussed here: https://github.com/Roasbeef/lightning-rfc/pull/1#discussion_r886414660

Reviewer 4: the script now in the bolt does the same thing as your script, the difference is the on-chain footprint (weight units)

Reviewer 5: I understand it functionally completes the same; I strongly suspect 1 WU is not worth the other costs, such as readability, analysis, miniscript compatbility(which is related).

It's a really interesting trick though!

Author: Yeah we understand that the behavior is the same, the discussion here now is if we should just have that extra byte which makes things much easier to understand at a glance (this has come up several times in code review so far in https://github.com/lightningnetwork/lnd/pulls).

Reviewer 7: Not sure what decision was made here but I think it would make much more sense to save the on-chain space and be compatible with miniscript

        ```

In order to spend a offered HTLC, via either script path, an `inclusion_proof`
must be specified along with the control block. This `inclusion_proof` is
simply the `tap_leaf` hash of the path _not_ taken.

#### Accepted HTLCs

Accepted HTLCs inherit a similar format:

  * `OP_1 accepted_htlc_key`
  * where:
    * `accepted_htlc_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
    * `htlc_script_root = tapscript_root([htlc_timeout, htlc_success])`
    * `htlc_timeout`:
        ```
        <remote_htlcpubkey> OP_CHECKSIG
        OP_CHECKSEQUENCEVERIFY
        <cltv_expiry> OP_CHECKLOCKTIMEVERIFY OP_DROP
        ```
    * `htlc_success`:
        ```
        OP_SIZE 32 OP_EQUALVERIFY OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY
        <local_htlcpubkey> OP_CHECKSIGVERIFY

>         <remote_htlcpubkey> OP_CHECKSIG

Reviewer 6: Missing a  OP_CHECKSEQUENCEVERIFY, same as in the Offered HTLC case (timeout path)

Reviewer 4: Not needed because the remote signature prevents you from changing it

Reviewer 6: hmm not sure but we are talking anchor ANYONE_CANPAY signatures ? For Static Remote you are right but for Anchor-Channels its needed.

Reviewer 6: see also [option anchors](https://github.com/lightning/bolts/blob/master/03-transactions.md#offered-htlc-outputs)

Reviewer 6: As @antonilol said, its not needed because the signature guards the sequence of the followup transaction.

    previously received in `funding_created` using `NonceAgg` from `bip-musig2`.
  
  - MUST use the generated secret nonce and the calculated aggregate nonce to
    construct a `musig2` partial signature for the sender's remote commitment
    using the `Sign` algorithm from `bip-musig2`.
  
  - MUST include the partial signature and the public counterpart of the
    generated nonce in the `partial_signature_with_nonce` field.

The recipient:

  - MUST fail the channel if `partial_signature_with_nonce` is absent.
  
  - MUST compute the aggregate nonce from:
  
    - the `next_local_nonce` field the recipient previously sent in
      `funding_created`
  
    - the `public_nonce` included as part of the `partial_signature_with_nonce`
      field
  
  - MUST verify the `partial_signature_with_nonce` field using the
    `PartialSigVerifyInternal` algorithm of `bip-musig2`:
  
    - if the partial signature is invalid, MUST fail the channel

#### `channel_ready` Extensions

We add a new TLV field to the `channel_ready` message:

1. `tlv_stream`: `channel_ready_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:
      * [`66*byte`: `public_nonce`]

> Similar to the `next_per_commitment_point`, by sending the `next_local_nonce`

Author: That's unchanged?

public key script template:
```
OP_1 <taproot_output_key>
```

The `taproot_output_key` is a 32-byte x-only `secp256k1` public key, with all
digital signatures based off of the schnorr signature scheme described in [BIP
340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki).

A `taproot_output_key` commits to an internal key, and optional script root via
the following mapping:
```
taproot_output_key = taproot_internal_key + tagged_hash("TapTweak", taproot_internal_key || script_root)*G
```

It's important to note that while `taproot_output_key` is serialized as a
32-byte public key, in order to properly spend the script path, the parity
(even or odd) of the output public key must be remembered.

The `taproot_internal_key` is also a BIP 340 public key, ideally derived anew
for each output. The `tagged_hash` scheme is described in BIP 340, we briefly
define the function as:
```
tagged_hash(tag, msg) = SHA256(SHA256(tag) || SHA256(tag) || msg)
```

#### Tapscript Tree Semantics

The `script_root` is the root of a Tapscript tree as defined in BIP 341. A tree
is composed of `tap_leaves` and `tap_branches`.

> A `tap_leaf` is a two tuple of (`leaf_version`, `leaf_script`). A `tap_leaf` is serialized as:

Reviewer: i wonder if the correctness of the `leaf_version` (i.e == `0xc0` or == `0xc1`) is verified once the funding output has hit the chain (no `option_zero_conf` mode) to be sure the `script_root` built from counterparty contributions  is the correct one committed on-chain.

from looking on `channel_ready` recipient, there is no say that the `script_root` must be verified. it might be elsewhere in this document. e.g for offered HTLCs there is a split of the timeout and success paths in individual script leaves.

     sign the commitment transaction.

  2. Each commitment state has a unique number which is currently encoded as a
     48-bit integer across the sequence and lock time of the commitment
     transaction. 

  3. The shachain scheme is used today to generate a fresh nonce-like value for
     the revocation scheme of today's penalty based channels.

  4. A verification nonce is only used to _sign_ a local party's commitment
     transaction when they go to broadcast a force close transaction.

  5. The remote party never sees the unaggregated partial signature the local
     party will use to broadcast, only the aggregated signature is seen on chain.

  6. As a result of all the above, a verification nonce will never be used to
     sign the exact same commitment transaction, thereby avoiding nonce reuse.

As a result, we recommend the following counter based scheme to generate
verification nonce. With this scheme, signing can remain stateless, as given
the commitment height/number of the state to sign, and the original shachain
root, the verification nonce sent by that state can be deterministically
reproduced:  

>   1. Given the shachain root used to generate revocation pre-images,

Reviewer 5: has any thought been given on how this handles/could be extended to handle splicing? 

you're going to possibly be signing multiple times at a given height, and sharing nonces before knowing the final txid/state of the commit tx.

Without giving much thought the splices could be given a counter and incorporated to the shachain scheme.

cc @ddustin

be taken.

We refer to the `simple_taproot_nums` as the following value:
```
02dca094751109d0bd055d03565874e8276dd53e926b44e3bd1bb6bf4bc130a279
```

The value was [generated using this
tool](https://github.com/lightninglabs/lightning-node-connect/tree/master/mailbox/numsgen),
with the seed phrase "Lightning Simple Taproot".

## Design Overview

With the preliminaries out of the way, we provide a brief overview of the
Simple Taproot Channel design.

The multisig output becomes a single P2TR key, with `musig2` key aggregation
used to combine two keys into one.

For commitment transactions, we inherit the anchor output semantics, meaning
there are two outputs used for CPFP, with all other scripts inheriting a `1
CSV` restriction.

The local output of the commitment transaction uses a script-path based
revocation scheme in order to ensure that the information needed by 3rd parties
to sweep the anchor outputs is always revealed on chain.

The remote output of the commitment transaction uses the `combined_funding_key`
as the top-level internal key, and then commits to a normal remote script.

Anchor outputs use the `local_delayedpubkey` and the `remotepubkey` of both

> parries as the top-level

Reviewer 8: ```suggestion
parties as the top-level
```

### New TLV Types

Note that these TLV types exist across different messages, but their type IDs are always the same.

#### partial_signature_with_nonce
- type: 2
- data:
   * [`32*byte`: `partial_signature`]
   * [`66*byte`: `public_nonce`]

#### next_local_nonce
- type: 4
- data:
   * [`66*byte`: `public_nonce`]

#### partial_signature
- type: 6
- data:
   * [`32*byte`: `partial_signature`]

#### shutdown_nonce
- type: 8
- data:
   * [`66*byte`: `public_nonce`]

### Channel Funding

`n_a_L`: Alice's local secret nonce

`n_B_R`: Bob's remote public nonce

![](taproot_channel_open.jpg)

The `open_channel` and `accept_channel` messages are extended to specify a new
TLV type that houses the `musig2` public nonces.

We add `option_simple_taproot` to the set of defined channel types.

#### `open_channel` Extensions

1. `tlv_stream`: `open_channel_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:
      * [`66*byte`:`public_nonce`]

##### Requirements

The sending node:

  - MUST specify the `next_local_nonce` field.
  - MUST use the `NonceGen` algorithm defined in `bip-musig2` to generate
    `next_local_nonce` to ensure it generates nonces in a safe manner.

>   - MOST not set the `announce_channel` bit.

Reviewer 9: ```suggestion
  - MUST not set the `announce_channel` bit.
```

    points

##### Rationale

For the initial Taproot channel type (`option_simple_taproot`), musig2 nonces
must be exchanged in the first round trip (open->, <-accept). This is done to
ensure to ensure that the initiator is able to generate a valid musig2 partial
signature at the next step once the transaction to be signed is fully
specified.

The `next_local_nonce` field will be used to verify incoming commitment
signatures. Each incoming signature will carry the newly generated nonce used
to sign the commitment transaction. At force close broadcast time, the
verification nonce is then used to sign the local party's commitment
transaction.

#### `accept_channel` Extensions

1. `tlv_stream`: `accept_channel_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:
      * [`66*byte`:`public_nonce`]

##### Requirements

The sender:

  - MUST set `next_local_nonce` to the `musig2` public nonce used to sign local
    commitments as specified by the `NonceGen` algorithm of `bip-musig2`.

The recipient:

  - MUST reject the channel if `next_local_nonce` is absent, or cannot be
    parsed as two compressed secp256k1 points

>   - the specified public nonce cannot be parsed as two compressed secp256k1

Reviewer 9: ```suggestion
```
This is redundant with the above MUST clause

The recipient:

  - MUST fail the channel if `partial_signature_with_nonce` is absent.
  
  - MUST compute the aggregate nonce from:
  
    - the `next_local_nonce` field the recipient previously sent in
      `open_channel`
  
    - the `public_nonce` included as part of the `partial_signature_with_nonce`
      field
  
  - MUST verify the `partial_signature_with_nonce` field using the
    `PartialSigVerifyInternal` algorithm of `bip-musig2`:
  
    - if the partial signature is invalid, MUST fail the channel

#### `channel_ready` Extensions

We add a new TLV field to the `channel_ready` message:

1. `tlv_stream`: `channel_ready_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:
      * [`66*byte`: `public_nonce`]

Similar to the `next_per_commitment_point`, by sending the `next_local_nonce`
value in this message, we ensure that the remote party has our public nonce
which is required to generate a new commitment signature.

##### Requirements

The sender:

- MUST set `next_local_nonce` to a fresh, unique `musig2` nonce as specified by
  `bip-musig2`

The recipient:

- MUST fail the channel if `next_local_nonce` is absent.

### Cooperative Closure

> Compared to the base segwit v0 channel type, for simple taproot channels, then

Reviewer 9: ```suggestion
Compared to the base segwit v0 channel type, for simple taproot channels, the
```

co-op close transaction now always signals RBF. In other words, the sequence
field of the sole input to the cooperative close transaction MUST be
less-than-or-equal to `0xfffffffd`. This enables a future cooperative closure
flow to support increasing the fee of subsequent close offers via RBF.

In addition, rather than adopt the existing cooperative closure fee rate
"negotiation", the responder SHOULD now always accept the offer sent by the
initiator. In other words, the cooperative close process now terminates after
exactly 1 RTTs: initiator sends sigs with offer, with the responder echo'ing
back the same fee rate. This serves to ensure that the co-op close process
always terminates deterministically, and also plays nicer with the nonces: only
a single message is ever signed by both sides for a coop close workflow.

#### `shutdown` Extensions

A new TLV field is added to the `shutdown` message:

1. type: 38 (`shutdown`)
2. data:
    * ...
    * [`shutdown_tlvs`:`tlvs`]

1. `tlv_stream`: `shutdown_tlvs`
2. types:
    1. type: 8 (`shutdown_nonce`)

>     2: data:

Reviewer 9: ```suggestion
    2. data:
```

        * [`66*byte`:`nonces`]

Before a signature can be generated for a co-op close transaction, both sides
must exchange a fresh pair of `musig2_pubnonce` values. We package this with
the `shutdown` message so that both sides can send a `closing_signed` message
as soon as a `shutdown` message is sent and received.

Only a single nonce is needed as there's only a single message to sign: the
shared co-op close transaction.

##### Requirements

A sending node:

 - MUST set the `shutdown_nonce` to a valid `musig2` public nonce.

A receiving node:

 - MUST verify that the `shutdown_nonce` value is a valid `musig2` public
   nonce.

#### `closing_signed` Extensions

We add a new TLV to the `closing_signed`
message's existing `tlv_stream`:

1. `tlv_stream`: `closing_signed_tlvs`
2. types:
   1. type: 6 (`partial_signature`)
   2. data:

>       * [`32*byte`: `partial_signature`

Reviewer 9: ```suggestion
      * [`32*byte`: `partial_signature`]
```

    [BIP 342](https://github.com/bitcoin/bips/blob/master/bip-0342.mediawiki), as
    a BIP 340 Schnorr signature (non-partial).

    - Each HTLC signature must be packed as a 64-byte byte value within the
      existing `htlc_signature` field.

The recipient:

  - MUST fail the channel if `signature` is non-zero.

  - MUST fail the channel if `partial_signature_with_nonce` is absent.

  - MUST compute the aggregate nonce from:

    - the `next_local_nonce` field the recipient previously sent in the latest
      of `channel_ready`, `channel_reestablish`, or `revoke_and_ack`

    - the `public_nonce` included as part of the `partial_signature_with_nonce`
      field

  - MUST verify the `partial_signature_with_nonce` field using the
    `PartialSigVerifyInternal` algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

  - MUST fail the channel if _any_ of the received HTLC signatures is invalid
    according to BIP 342.

#### `revoke_and_ack` Extensions

A new TLV stream is added to the `revoke_and_ack` message:

1. `tlv_stream`: `revoke_and_ack_tlvs`
2. types:
   1. type: 4 (`next_local_nonce`)
   2. data:

>       * [`66*byte`: `public_nonce`]

Reviewer 9: The [counter based nonce generation](https://github.com/lightning/bolts/blob/e25132d8de0164224578964fcd3f7328ddfc3281/bolt-simple-taproot.md#counter-based-nonce-generation--jit-nonce-strengthening) scheme recommends using the commitment index encoded as a 48-bit integer across the sequence and lock time of the commitment transaction to avoid having to store nonces.

Each splice commitment would be at the same commitment state, but have a different funding tx. Does this mean different splice commitments would use the same set of verification public nonces? 

My understanding is that reusing the same nonce and key for different messages would be unsafe. If that is the case, then we might need to rethink the indexing scheme to leave room for unique indexes for each pending splice commitment with the same commitment index.

>       * [`66*byte`: `public_nonce`]

Author: > We still have a single revoke_and_ack message that revokes those N commitments at once, because they still share the same revocation secret.

Interesting, for some reason I thought maybe they'd all have different revocation secrets, so you'd still send N RAA messages, but this makes sense. 

> To be able to generate signatures for the N next commitments, revoke_and_ack must then contain N next_local_nonce, right?

That's right, the signer needs N nonces from the other peer to be able to generate N new commitment sigs for the next state. 

Is there a hard upper limit on the number of pending splices permitted?

>       * [`66*byte`: `public_nonce`]

Author: > Each splice commitment would be at the same commitment state, but have a different funding tx. Does this mean different splice commitments would use the same set of verification public nonces?

Not factoring in the message comment above, yes. IIUC, each splice transaction also has its own protocol-specific sequence number. If this is encoded in the splice kick off transaction, then you can use this as another dimension if generating the nonces with the counter based scheme. So it'd be `sha_chain.leaf(spliceNo || commitNo)`. 

> My understanding is that reusing the same nonce and key for different messages would be unsafe.

Correct! I think w/ the above we can just use the splice sequence number as another dimension for each signed commitment?

>       * [`66*byte`: `public_nonce`]

Reviewer 9: That sounds like a reasonable approach.

The `commitNo` must already be stored for each commit, but implementers would need to just store the `spliceNo` for the last splice confirmed on-chain until the funding tx of a new splice replaces it.

>       * [`66*byte`: `public_nonce`]

Reviewer 10: Agreed, there is an ordering on splice transactions, we can use that to generate distinct nonces using the shachain, that shouldn't be an issue.

> Is there a hard upper limit on the number of pending splices permitted?

Not yet, but there should be one, because we cannot send an unbounded number of nonces in `revoke_and_ack` (since message size is limited). In practice the number of pending splices is bounded though, because nodes require the feerate to increase across splices.

By using the musig2 sorting routine, we inherit the lexicographical key
ordering from the base segwit channels. Throughout the lifetime of a channel
any keys aggregated via musig2 also are assumed to use the `KeySort` routine

### Commitment Transactions

#### To Local Outputs

For the simple taproot commitment type, we use the same flow of revocation or
delay, while re-using the NUMS point to ensure that the internal key is always
revealed (script path always taken). This ensures that the anchor outputs can
_always_ be spent given chain revealed information. For the remote party, the
`remotepubkey` will be revealed once the remote party sweeps. For the local
party, we ensure that the `local_delayedpubkey` is revealed with an extra noop
data push.

The new output has the following form:

  * `OP_1 to_local_output_key`
  * where:
    * `to_local_output_key = taproot_nums_point + tagged_hash("TapTweak", taproot_nums_point || to_delay_script_root)*G`
    * `to_delay_script_root = tapscript_root([to_delay_script, revoke_script])`
    * `to_delay_script` is the delay script:
        ```
        <local_delayedpubkey> OP_CHECKSIG
        <to_self_delay> OP_CHECKSEQUENCEVERIFY OP_DROP

>     * `revoke_script` is the delay script:

Reviewer 5: ?
```suggestion
    * `revoke_script` is the revoke script:
```

>         ```

Author: What do you mean? Looking at the other changes for miniscript compat, we just need to make the `OP_CHECKSIG` into an `OP_CHECKSIGVERIFY`?

>         ```

Author: Ah that makes sense. Will leave this one as is for now... 

In an alternate dimension, we would've used the annex here, as we just want to make sure it's published so others can use this value to sweep the anchors.

        <local_delayedpubkey> OP_DROP
        <revocation_pubkey> OP_CHECKSIG
        ```

The parity (even or odd) of the y-coordinate of the derived
`to_local_output_key` MUST be known in order to derive a valid control block.

The `tapscript_root` routine constructs a valid taproot commitment according to
BIP 341+342. Namely, a leaf version of `0xc0` MUST be used. 

In the case of a commitment breach, the `to_delay_script_root` can be used
along side `<revocationpubkey>` to derive the private key needed to sweep the
top-level key spend path. The control block can be crafted as such:
```

> revoke_control_block = (output_key_y_parity | 0xc0) || taproot_nums_point || revoke_script

Author: Good point, would we prefer that the derivation of the control block is more implicit in the spec? I think it's relevant at time to spell it out, as the internal key isn't always `taproot_nums_point`. So maybe we just need to specify what internal key should be used, then refer to the section you've linked.

> revoke_control_block = (output_key_y_parity | 0xc0) || taproot_nums_point || revoke_script

Reviewer 11: Yes I think that specifying the script and internal key used to build the proof that must be included in the control block should be enough.

```

In the case of a routine force close, the script path must be revealed so the
broadcaster can sweep their funds after a delay. The control block to spend is
only `33` bytes, as it just includes the internal key (along with the y-parity
bit and leaf version):
``` 
delay_control_block = (output_key_y_parity | 0xc0) || taproot_nums-point || to_delay_srcipt
```

A valid witness is then:
```
<local_delayedsig> <to_delay_script> <delay_control_block>
```

As with base channels, the `nSequence` field must be set to `to_self_delay`.

#### To Remote Outputs

As we inherit the anchor output, semantics we want to ensure that the remote
party can unilaterally sweep their funds after the 1 block CSV delay. In order
to achieve this property, we'll utilize a NUMS point (`simple_taproot_nums`) By
using this point as the internal key, we ensure that the remote party isn't
able to by pass the CSV delay.

Using a NUMS key has a key benefit: the static internal key allows the remote
party to scan for their output on chain, which is useful for various recovery
scenarios.

The `to_remote` output has the following form:

  * `OP_1 to_remote_output_key`
  * where:

>     * `taproot_nums_point = 0245b18183a06ee58228f07d9716f0f121cd194e4d924b037522503a7160432f15`

Author: Correct! The linked point matches what we use in lnd today: https://github.com/lightningnetwork/lnd/blob/2f2efc782436944ce98b1e0e13bde125951b2b36/input/script_utils.go#L46-L48

Can't recall why the text has this incorrect point. 

As a side question @sstone, here's the code we used to generate the NUMs point: https://github.com/lightninglabs/lightning-node-connect/blob/master/mailbox/numsgen/main.go. Perhaps Eclair is interested in replicating the derivation in Scala/Kotlin to increase confidence in the integrity of the point?

```
<remote_sig> <to_remote_script> <to_remote_control_block>
```

where `to_remote_control_block` is:
```
(output_key_y_parity | 0xc0) || combined_funding_key
```

The `sequence` field of the input MUST also be set to `1`. 

#### Anchor Outputs

For simple taproot channels (`option_simple_taproot`) we'll maintain the same
form as base segwit channels, but instead will utilize `local_delayedpubkey`
and the `remotepubkey` rather than the multi-sig keys as that's no longer
revealed due to musig2.

An anchor output has the following form:

  * `OP_1 anchor_output_key`
  * where:
    * `anchor_internal_key = remotepubkey/local_delayedpubkey`
    * `anchor_output_key = anchor_internal_key + tagged_hash("TapTweak", anchor_internal_key || anchor_script_root)`
    * `anchor_script_root = tapscript_root([anchor_script])`
    * `anchor_script`:
        ```
        OP_16 OP_CHECKSEQUENCEVERIFY
        ```

This output can be swept by anyone after 16 blocks with the following witness:
```
<anchor_script> <anchor_control_block>
```

where `anchor_control_block` is:
```
(output_key_y_parity | 0xc0) || anchor_internal_key
```

`nSequence` needs to be set to 16.

> ### HTLC Scripts & Transactions

Reviewer 10: Did you check compatibility with miniscript? The existing (non-taproot) scripts we use diverge from what the miniscript compiler generates, which is very annoying. Since we have an opportunity to slightly change the scripts we use, I would really like them to be miniscript-compatible, which allows importing the miniscript policies for a forced-close channel into any miniscript-compatible wallet (e.g. bitcoin core), which saves one transaction (the 2nd-level transaction) and thus saves on-chain fees.

@sstone is working on this on our side and can help with the miniscript policies.

Author: > Did you check compatibility with miniscript? 

IIRC someone did pretty early on. We had a script that was slightly more optimized, but harder to reason about for humans, so we used a simpler variant. Ultimately, all our scripts are still pretty simple.

I think miniscript compatibility is a laudable goal. FWIW we don't really use it much on the `lnd` side, as we didn't really see a use case as we don't use the `bitcoind` wallet, and subjectively there hasn't been much widespread ecosystem uptake yet (eg: I don't think there's a BIP yet?).

IIUC, only recently has miniscript also been updated to express taproot script trees? 

> The existing (non-taproot) scripts we use diverge from what the miniscript compiler generates, which is very annoying. 

Re the import case you brought up above (I'm not very familiar with this tooling), does this mean that if a script isn't identical to one that would've originally been generated by miniscript, that it can't be import into the `bitcoind` wallet? I thought maybe there'd be some lower level representation (just the script with some additional sugar) that could be used.

Reviewer 12: Re the import case: yeah you need the Miniscript descriptor to precisely describe the script, otherwise you won't find the corresponding coins. I think there may be some confusion with the "Miniscript policy" language. You can import a Miniscript descriptor which has the same semantics as, say, an HTLC output script but you would only find the coins associated with it if you import the script corresponding to this precise implementation of the HTLC semantic.

Reviewer 11: > Re the import case: yeah you need the Miniscript descriptor to precisely describe the script, otherwise you won't find the corresponding coins. I think there may be some confusion with the "Miniscript policy" language. You can import a Miniscript descriptor which has the same semantics as, say, an HTLC output script but you would only find the coins associated with it if you import the script corresponding to this precise implementation of the HTLC semantic.

With the current LN scripts, all we can do is import raw scripts/addresses: bitcoin core won't be aware of any CSV delays and won't know how to spend them.
But our understanding is that, with the ongoing work on descriptors/miniscript/tapscript, if our LN scripts are written in miniscript then every time a channel is force-closed and a commit tx published, we could import a new descriptor that describes our outputs in the commit tx (at least our `to-local` or `to-remote` output) and bitcoin core would know when these outputs become spendable and how to spend them. Once these outputs have been spent then these descriptors could be remove from the wallet (though I believe that is not supported yet).

value in this message, we ensure that the remote party has our public nonce
which is required to generate a new commitment signature.

##### Requirements

The sender:

- MUST set `next_local_nonce` to a fresh, unique `musig2` nonce as specified by
  `bip-musig2`

The recipient:

- MUST fail the channel if `next_local_nonce` is absent.

### Cooperative Closure

Compared to the base segwit v0 channel type, for simple taproot channels, the
co-op close transaction now always signals RBF. In other words, the sequence
field of the sole input to the cooperative close transaction MUST be
less-than-or-equal to `0xfffffffd`. This enables a future cooperative closure
flow to support increasing the fee of subsequent close offers via RBF.

In addition, rather than adopt the existing cooperative closure fee rate
"negotiation", the responder SHOULD now always accept the offer sent by the
initiator. In other words, the cooperative close process now terminates after
exactly 1 RTTs: initiator sends sigs with offer, with the responder echo'ing
back the same fee rate. This serves to ensure that the co-op close process
always terminates deterministically, and also plays nicer with the nonces: only

> a single message is ever signed by both sides for a coop close workflow.

Reviewer 10: Looks like I missed that paragraph in my commits, which refers to the legacy closing protocol. Now that we updated this spec to rely on `option_simple_close` whenever using official taproot channels, we should remove this paragraph?

It looks like we can actually merge this section and the "Channel Cooperative Close Extensions" below into a single section and remove references to legacy vs modern close entirely?

#### `shutdown` Extensions

For taproot channels, the shutdown message includes a single nonce:

1. type: 38 (`shutdown`)
2. data:
    * ...
    * [`shutdown_tlvs`:`tlvs`]

1. `tlv_stream`: `shutdown_tlvs`
2. types:
    1. type: 8 (`shutdown_nonce`)
    2. data:
        * [`66*byte`:`public_nonce`]

The `shutdown_nonce` represents the sender's "closee nonce" - the nonce they
will use when sending `closing_sig` messages. This applies to both the legacy
`closing_signed` flow and the modern RBF flow using

> `closing_complete`/`closing_sig`.

Reviewer 10: I missed that one as well :/

```suggestion
will use when sending `closing_sig` messages.
```

- `closing_sig` uses `PartialSig` (32 bytes) with a separate `NextCloseeNonce`
  field

With this pattern, we exchange the set of `closee_nonces` up front in the
`shutdown` message. When either party wants to initiate a new update, they send
`closing_complete` which includes their JIT `closer_nonce`. Sending this
signatures serves to consume the `closee_nonce` of the remote party, so then
they respond with `closing_sig`, they include a new `next_closee_nonce`.

#### `closing_complete` Extensions

For taproot channels, the `closing_complete` message uses `PartialSigWithNonce`
to support MuSig2 signatures with JIT nonce delivery:

1. `tlv_stream`: `closing_tlvs`
2. types:
    1. type: 5 (`closer_no_closee`)
    2. data:
        * [`98*byte`:`partial_sig_with_nonce`]
    1. type: 6 (`no_closer_closee`)
    2. data:
        * [`98*byte`:`partial_sig_with_nonce`]
    1. type: 7 (`closer_and_closee`)
    2. data:
        * [`98*byte`:`partial_sig_with_nonce`]

Note: The TLV type numbers (5, 6, 7) are different from the non-taproot types
(1, 2, 3) to distinguish taproot signatures. Each `partial_sig_with_nonce`
contains:
- 32 bytes: MuSig2 partial signature

> - 66 bytes: The sender's next closee nonce for potential RBF iterations

Reviewer 11: Same here, what about `The sender's  closee nonce used to generate the partial signature` ?

##### Requirements

The sender of `closing_complete` (aka. "the closer"):

  - For taproot channels (when `option_simple_taproot` was negotiated):
    - MUST provide `PartialSigWithNonce` (98 bytes) in the appropriate TLV
      field based on output presence
    - MUST compute the aggregated `musig2` public key from the sorted
      `funding_pubkey`s using the `KeyAgg` algorithm from `bip-musig2`
    - For the first closing transaction:
      - MUST use the `shutdown_nonce` received from the peer as their closee
        nonce
      - MUST use a locally generated nonce as the closer nonce
    - For RBF iterations:
      - MUST use the nonce extracted from the peer's previous `closing_sig`
        message as their closee nonce
      - MUST use a fresh locally generated nonce as the closer nonce
    - MUST generate the partial signature using the `Sign` algorithm from
      `bip-musig2`

The receiver of `closing_complete` (aka. "the closee"):

  - For taproot channels:
    - MUST verify that signature fields contain `PartialSigWithNonce` (98
      bytes)
    - MUST extract the partial signature (first 32 bytes) and the sender's next

>       closee nonce (remaining 66 bytes)

Author: Here the distinct terminology was used on purpose. I used closee/closer here, as the roles are always non ambiguous. If you're sending `closing_complete`, then you're the closer. 

I think we should agree on what terminology we'll use here. I adopted the closee/closer based on the initial comment @t-bast left outlining a protocol (adapted it a bit as you see here).

    - MUST use the extracted nonce for verification (along with their own
      closee nonce)
    - MUST compute the aggregate nonce from:
      - the `shutdown_nonce` the recipient previously sent (their closee nonce)
      - the appropriate nonce for the closer role (from shutdown or previous
        RBF round)
    - MUST verify the partial signature using the `PartialSigVerifyInternal`
      algorithm of `bip-musig2`
    - MUST store (in memory) the extracted nonce for potential future RBF

>     iterations

Author: Agree, will remove.

        },
        {
          "remote_partial_sig_hex": "0967aa040669318987029fba55410f9a43c4dd57da9e3a60e74b83a35abea76e813fbbdaa8baae38e096a6b7718d75838cb33e264814678823f55ee6fb9cdd0f",
          "resolution_tx_hex": "02000000000101ec4c0a34c981864f9badcb8383bbe42ec6b32e68c2aa1a7c7c2e8422adde673f06000000000100000001a00f000000000000225120df20bcec43daa75161f7d013254e401812e0fee8bc3369220b6a33672fc18ba0054181dd0918b0e01f4c1f701a5689f3c5c076bf71d4f365f21b67707527c9e71fe40e0c6fee2f3e46ee297d542e883f5ba5004b53b04cda060b6f99d3a8b2c3f48883409c3edb75f915ab9573cd44d9500e630fb6c2388c21733be54bce3824e66df526bf73df2206ccb8fbd125a4f0025e5bae8ecd7ea586d62ff341cead78c0e2528b0020000000000000000000000000000000000000000000000000000000000000000041c1d4c77088d346bce67c13bbbf82ca112588f4b1c9595a1f8af3be9b2f95a109a06c3390c812b2596986592f02c7f22e4f857fb553805ff9ac1c2bda361c47c3fb00000000"
        }
      ]
    },
    {

>       "name": "commitment tx with some HTLCs trimmed",

Author: Should be fixed in an upcoming push

    * sequence: 1
    * witness: `<remotehtlcsig> <localhtlcsig> <preimage> <htlc_success_script> <control_block>`
  * output:
    * value: htlc_value
    * script:
      * OP_1 htlc_success_key
      * where:
        * `htlc_success_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
        * `htlc_script_root = tapscript_root([htlc_success])`
        * `htlc_success`:
        ```
        <local_delayedpubkey> OP_CHECKSIGVERIFY
        <to_self_delay> OP_CHECKSEQUENCEVERIFY
        ```

##### HTLC-Timeout Transactions

A HTLC-Timeout transaction has the following structure:

  * version: 2
  * locktime: cltv_expiry
  * input:
    * txid: commitment_tx
    * vout: htlc_index
    * sequence: 1
    * witness: `<remotehtlcsig> <localhtlcsig> <htlc_timeout_script> <control_block>`
  * output:
    * value: htlc_value
    * script:
      * OP_1 htlc_timeout_key
      * where:
        * `htlc_timeout_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
        * `htlc_script_root = tapscript_root([htlc_timeout])`
        * `htlc_timeout`:
        ```
        <local_delayedpubkey> OP_CHECKSIG

>         <to_self_delay> OP_CHECKSEQUENCEVERIFY OP_DROP

Reviewer 13: The prod tapscript variants that use `OP_CHECKSIGVERIFY` instead of `OP_CHECKSIG` ... `OP_DROP` leave the time-lock value as the final (and only) stack element. Script success then depends on that value being non-zero (since 0 is falsy in Bitcoin script).

This affects three scripts:

  - to_delay_script: `<key> OP_CHECKSIGVERIFY <to_self_delay> OP_CHECKSEQUENCEVERIFY` - final stack is [`to_self_delay`]                                                                          
  - Accepted HTLC timeout: `<key> OP_CHECKSIGVERIFY 1 OP_CSV OP_VERIFY <cltv_expiry> OP_CLTV` - final stack is [`cltv_expiry`]                                                                    
  - 2nd-level HTLC outputs: `<key> OP_CHECKSIGVERIFY <delay> OP_CSV` - final stack is [`delay`]

Note that to_remote_script is not affected since it hardcodes 1 `OP_CHECKSEQUENCEVERIFY`, which always leaves a truthy [1] on the stack.                                                      

In practice this is safe, since `to_self_delay` must be positive per BOLT 2 negotiation, CLTV expiries are always future block heights, and CSV delays are always > 0. But unlike the legacy `OP_CHECKSIG` ... `OP_DROP` scripts where the final stack element was always the signature check result (1), these scripts have an implicit invariant that the time-lock parameters must be non-zero for the script to succeed.                                                                                                                                                         
                  
  Should we add a brief note in the spec calling out this invariant? Something like:                                                                                                          
   
> Note: because these scripts use OP_CHECKSIGVERIFY (which consumes the boolean result) followed by a terminal OP_CHECKSEQUENCEVERIFY or OP_CHECKLOCKTIMEVERIFY (which leaves its argument on 
  the stack), the time-lock value serves as the final truthy stack element. Implementations MUST ensure these values are non-zero.

 There is also a minor specification consistency issue previously pointed out by @sstone and currently under review in an LND PR by @gijswijs. The HTLC-Timeout second-level output description (under “HTLC-Timeout Transactions”) still shows the legacy `OP_CHECKSIG` … `OP_DROP` pattern, while the test vectors use `OP_CHECKSIGVERIFY` for both success and timeout second-level outputs. The prose likely needs updating to match the test vectors.

Reviewer 10: > Should we add a brief note in the spec calling out this invariant?

I think this is already explained in the accepted HTLCs section (https://github.com/Roasbeef/lightning-rfc/blob/simple-taproot-chans/bolt-simple-taproot.md#accepted-htlcs)?

Reviewer 13: > I think this is already explained in the accepted HTLCs section (https://github.com/Roasbeef/lightning-rfc/blob/simple-taproot-chans/bolt-simple-taproot.md#accepted-htlcs)?

That note explains why `CSV` is omitted from certain HTLC paths. It doesn't address the stack semantics point. The broader issue is that across all three `CHECKSIGVERIFY` + terminal timelock scripts, the timelock value itself ends up as the final stack element, so script success implicitly requires it to be non-zero. That invariant isn't called out anywhere currently.

It's not strictly necessary since these values are always non-zero in practice, but a brief note would help readers who are tracing the stack logic and wondering what the final truthy element is.

|-----|-------|
| `local_funding_privkey` | `"local-funding"` |
| `remote_funding_privkey` | `"remote-funding"` |
| `local_payment_basepoint_secret` | `"local-payment-basepoint"` |
| `remote_payment_basepoint_secret` | `"remote-payment-basepoint"` |
| `local_delayed_payment_basepoint_secret` | `"local-delayed-payment-basepoint"` |
| `remote_revocation_basepoint_secret` | `"remote-revocation-basepoint"` |
| `local_htlc_basepoint_secret` | `"local-htlc-basepoint"` |
| `remote_htlc_basepoint_secret` | `"remote-htlc-basepoint"` |
| `local_per_commit_secret` | `"local-per-commit-secret"` |

### MuSig2 Nonce Generation

MuSig2 partial signatures require deterministic nonces so that test vectors are
reproducible across implementations. The signing randomness for each party's
JIT nonces is derived as `SHA256(seed || label)` using:

| Party | Label |
|-------|-------|
| Local signer | `"local-signing-rand"` |
| Remote signer | `"remote-signing-rand"` |

The resulting 32-byte hash is used as the nonce randomness input to the MuSig2
signing session. Each transaction test case includes `local_nonce` and
`remote_nonce` fields (66-byte compressed public nonces) so implementations can

> verify the nonces were derived correctly and reconstruct the combined signature.

Reviewer 10: I couldn't figure out what you mean here. I've been unable to generate nonces that match yours. I've tried using `secp256k1_musig_nonce_gen` (see https://github.com/bitcoin-core/secp256k1/blob/1aafe15139976b0142d791aaf4963de3fc1ff736/examples/musig.c#L128) with `session_secrand` set to `SHA256(hex"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f" || "local-signing-rand".getBytes(ASCII))` but it doesn't match. I've tried omitting the `msg32` input and including it (set to the commit tx's `txid`) but it doesn't match either.

The fact that your nonces are different for each test vector (while the funding and commitment keys are the same) seem to imply that this isn't just deterministically derived based on the static seed and this label, there must be something related to the funding tx or commit tx that you're injecting as well?

I don't think we need to constrain how nonces are generated for those tests in the first place, it only creates more confusion. Can you simply include in each test vector the `local_secret_nonce` and `remote_secret_nonce` (and keep the existing `local_public_nonce` and `remote_public_nonce` as well)? That would be much easier for me to test 🙏

Author: > Can you simply include in each test vector the local_secret_nonce and remote_secret_nonce (and keep the existing local_public_nonce and remote_public_nonce as well)? That would be much easier for me to test 🙏

Done! 

Note it's a 97 byte nonce (64 bytes for the secret, 33 bytes public key).

Reviewer 10: I don't think you've pushed this fix?

```

It's important to note that while `taproot_output_key` is serialized as a
32-byte public key, in order to properly spend the script path, the parity
(even or odd) of the public key must be remembered.

The `taproot_internal_key` is also a BIP 340 public key, ideally derived anew
for each output. The `tagged_hash` scheme is described in BIP 340, we briefly
define the function as:
```
tagged_hash(tag, msg) = SHA256(SHA256(tag) || SHA256(tag) || msg)
```

#### Tapscript Tree Semantics

The `script_root` is the root of a Tapscript tree as defined in BIP 341. A tree
is composed of `tap_leaves` and `tap_branches`. 

A `tap_leaf` is a two tuple of (`leaf_version`, `leaf_script`). A `tap_leaf` is serialized as:
```
leaf_version || compact_size_of(leaf_script) || leaf_script
```
where `compact_size_of` is the variable length integer used in the Bitcoin p2p
protocol.

The digest of a `tap_leaf` is computed as:
```
tagged_hash("TapLeaf", leaf_encoding)
```

A `tap_branch` can commit to either a `tap_leaf` or `tap_branch`. Before
hashing, a `tap_branch` sorts the two `tap_node` arguments based on
lexicographical ordering:
```

> tagged_hash("TapBranch, sort(node_1, node_2)

Reviewer 14: ```suggestion
tagged_hash("TapBranch", sort(node_1, node_2)
```

Reviewer 4: ```diff
-tagged_hash("TapBranch, sort(node_1, node_2)
+tagged_hash("TapBranch", sort(node_1, node_2))
                                              ^
```

Reviewer 4: fixed in https://github.com/Roasbeef/lightning-rfc/pull/1

# Extension BOLT XX: Simple Taproot Channels

Authors:
  * Olaoluwa Osuntokun <roasbeef@lightning.engineering>

>   * Eugene Siegel <eugene@lightning.engineerg>

Reviewer 14: ```suggestion
  * Eugene Siegel <eugene@lightning.engineering>
```

  3. Next, each party generates a _partial_ signature and exchanges it with
    every other party.

  4. Finally, once all partial signatures are known, they can be combined into a
    valid schnorr signture, which can be veirfied using BIP 340.

Steps 1&2 may be performed out of order, or concurrently. In our case only two
parties exist, so as soon as one party knows both partial signatures, they can
be combined into a final signature.

#### Key Aggregation

We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)
for key aggregation. In order to avoid sending extra ordering information, we
always assume that both keys are sorted first with the `KeySort` algorithm
before aggregating (`KeyAgg(KeySort(p1, p2))`)>

#### Nonce Generation

A `musig2` public nonce is a 66-byte value which is the concentration of two EC
points encoded using the compressed format:
```
nonce_1 || nonce_2
```

Each public nonce has a corresponding _secret_ nonce which is only known to the
signer.

For nonce generation, we refer to the `NonceGen` algorithm defined in the
`musig2` BIP. We only specify the set of required arguments, though a signer

> can also opt to specify the additional argument sin order to strengthen their

Reviewer 5: ```suggestion
can also opt to specify the additional arguments in order to strengthen their
```

Author: Commenting here @instagibbs: AFAICT we're ok as far as nonce ordering, since it pairs up the "1st" and "2nd" nonces for each party, and EC addition is commutative.

nonce.

#### Nonce Handling

Throughput this document we always assume that _fresh_ nonces are generated for

> each signing attempt, and nonces are **never** written to disk. As we'll see

Author: Ideally yes...but there's a small quirk where for me to go force close, assuming I don't keep the "hot" signature on disk, I need my secret nonce to generate the signature. In theory, nodes could like encrypt the sig or w/e with some other key, so sort of key wrapping, with the assumption that one can only decrypt w/ that key.

> each signing attempt, and nonces are **never** written to disk. As we'll see

Author: Latest push adds a section above this describing a recommended counter based scheme to avoid persisting any nonces at all.

of time to reduce the number of round trips in the protocol.

Finally, due to the asymmetric state of the current revocation channels, we
actually require two sets of public nonces to be exchanged by each party: one
for the local commitment, and one for the remote commitment.

#### Signing

Once public nonces have been exchanged, the `NonceAgg` algorithm is used to
combine the public nonces into the combined nonce which will be a part of the
final combined signature.

The `Sign` is used to generate a valid partial signature, with the
`PartialSigVerify` algorithm used to verify each signature.

Once all partial signatures are known, the `PartialSigAgg` algorithm is used to
combine the partial signature into a final valid BIP 340 schnorr signature.

### Nothing Up My Sleeves Points

Whenever we want to ensure that a given P2TR can _only_ be spent via a script
path, we utilize a Nothing Up My Sleeve (NUMs) point. A NUMs point is an EC
point that no one knows the private key to. If no one knows the private key,
then it can't be used for key path signing, forcing the script path to always
be taken.

We refer to the `simple_taproot_nums` as the following value:
```

> 02dca094751109d0bd055d03565874e8276dd53e926b44e3bd1bb6bf4bc130a279

Reviewer 4: https://github.com/lightning/bolts/pull/995#issuecomment-1144469224

dont use a NUMS point but the revocation key for less fees and more privacy

> 02dca094751109d0bd055d03565874e8276dd53e926b44e3bd1bb6bf4bc130a279

Reviewer 4: https://github.com/lightning/bolts/pull/995#issuecomment-1167871874

or dont use a 1 csv locked to_remote output

> 02dca094751109d0bd055d03565874e8276dd53e926b44e3bd1bb6bf4bc130a279

Reviewer 15: you want a 1 csv to prevent the remote party seeing the commit tx in the mempool and pinning it

> 02dca094751109d0bd055d03565874e8276dd53e926b44e3bd1bb6bf4bc130a279

Reviewer 4: in my comment i explained how this would work, by removing the remote anchor and removing the 1 csv from the to_remote output the remote party still has 1 output to fee bump and pin if he wants just like with the anchor. pinning makes rbf hard, it hasnt anything to do with cpfp

> 02dca094751109d0bd055d03565874e8276dd53e926b44e3bd1bb6bf4bc130a279

Reviewer 15: Removing the anchor will work - I hadn't read your comment, so my comment can be ignored. But pinning does have to do with CPFP since it deals with the mempool and its package limits. However, I still don't think the change is worth adding more logic either at the spec or code level

> 02dca094751109d0bd055d03565874e8276dd53e926b44e3bd1bb6bf4bc130a279

Author: Yeh it's gone now in the latest diff. I agree with @Crypt-iQ not sure it's worth the trouble to re-work the anchors significantly. The latest diff re-uses the multi-sig key.

Anchor outputs use the respective local and remote keys from the commitment
transaction as the top level output, with the committed script being the `CSV
16` delay.

All HTLC scripts use the revocation key as the top-level internal key. This
allows the revocation to be executed without additional on-chain bytes, and
also reduces the amount of state nodes need to keep in order to properly handle
channel breaches.

## Specification

### Feature Bits

Inheriting the structure put forth in BOLT 9, we define a new feature bit to be
placed in the `init` message, `node_announcement` and also
`channel_announcement`. We opt to also place the bit in the
`channel_announcement` so taproot channels can be identified on the routing
level, which will eventually matter once PTLCs are fully rolled out.

| Bits  | Name                             | Description                                               | Context  | Dependencies      | Link                                  |
|-------|----------------------------------|-----------------------------------------------------------|----------|-------------------|---------------------------------------|

> | 50/51 | `option_simple_taproot`| Node supports simple taproot chanenls | INC | `option_channel_type`+`option_anchors` | TODO(roasbeef): link |

Author: Comment from @t-bast: can remove the channel announcement placement here.

Reviewer 10: I don't mind much if we keep it though if you think it has value, it's just that we'll have to define a different feature bit for PTLCs and it's the only one that will matter in `channel_announcement`

 * `C-`: presented in the `channel_announcement` message, but always odd (optional).
 * `C+`: presented in the `channel_announcement` message, but always even (required).
 * `9`: presented in [BOLT 11](11-payment-encoding.md) invoices.

The `option_simple_taproot` feature bit also becomes a defined channel type
feature bit for explicit channel negotiation.

Thought this document, we assume that `option_simple_taproot` was negotiated,
and also the `option_simple_taproot` channel type is used.

### Channel Funding

The `open_channel` and `accept_channel` messages are extended to specify a new
TLV type that houses the `musig2` public nonces:

We add `option_simple_taproot` to the set of defined channel types.

#### `open_channel` Extensions

1. `tlv_stream`: `open_channel_tlvs`
2. types:
    1. type: 2 (`local_musig2_pubnonce`)
    2: data:
        * [`66*byte`:`nonces`]
    1. type: 4 (`remote_musig2_pubnonce`)
    2: data:
        * [`66*byte`:`nonces`]

##### Requirements

The sending node:

  - MUST specify the `local_musig2_pubnonce` and `remote_musig2_pubnonce` TLV
    fields.

The sending node SHOULD:

  - use the `NonceGen` algorithm defined in `bip-musig2` to generate

>     `lcaol_musig2_pubnonce` and `remote_musig2_pubnonce` to  ensure it

Reviewer 4: fixed in https://github.com/Roasbeef/lightning-rfc/pull/1

    generates nonces in a safe manner.

The receiving node MUST fail the channel if:

  - the message doesn't include a `local_musig2_pubnonce` and
    `remote_musig2_pubnonce` value.

  - a specified public nonce cannot be parsed as two compressed secp256k1
    points

##### Rationale

For the initial Taproot channel type (`option_simple_taproot`), musig2 nonces
must be exchanged in the first round trip (open->, <-accept) to ensure that the
initiator is able to generate a valid musig2 partial signature at the next
step.

The `remote_musig2_pubnonce` will be used to sign the _remote_ party's
commitment transaction, while the `local_musig2_pubnonce` will be used to sign
the local party's commitment transaction. 

> We require two nonces, as there're actually two messages being signed: the

Author: Agreed, also looking for better terms than remote/local, as the meaning can differ depending on your "visualization PoV".

#### `accept_channel` Extensions

1. `tlv_stream`: `accept_channel_tlvs`
2. types:
    1. type: 2 (`local_musig2_pubnonce`)
    2: data:
        * [`66*byte`:`nonces`]
    1. type: 4 (`remote_musig2_pubnonce`)
    2: data:
        * [`66*byte`:`nonces`]

##### Requirements

The sender:

 - MUST set `local_musig2_pubnonce` and `remote_musig2_pubnonce` to the
   `musig2` public nonce as specified by the `NonceGen` algorithm of
   `bip-musig2`

 - Aggregate the received local+remote public nonces (into two combined nonces)
   using the `NonceAgg` algorithm of `bip-musig2`.

The receiver:

 - MUST reject the channel if `local_musig2_pubnonce` or
   `remote_musig2_pubnonce` are not present.

#### `funding_created` Extensions

The sender:

  - MUST derive the aggregated `musig2` public key using the exchanged `funding_pubkey`s 
    - Before aggregating the two funding public keys, the keys MUST be sorted
      using the `KeySort` routine of `bip-musig2`.

  - MUST use the exchanged `local_musig2_pubnonce` and `remote_musig2_pubnonce`
    received from the remote party with the `NonceAgg` algorithm to derive the
    combined nonce

>   - the `signature` field MUST be a `musig2` partial signature using the

Reviewer 16: just as a sanity check for myself, could you verify that this field is now only 32 bytes (in accordance with the `Sign` method's return type), and therefore the deserialization of the funding_created and funding_signed messages must depend on whether or not they're referring to Taproot channels?

Reviewer 16: Actually, reading your last paragraph:

> Re the partial sig serialization (need to catch up spec to this), I've opted to re-use the existing 64 byte sig field. Rather than just encoding the 32 byte s value, we'll also encode the combined nonce (R1||R2, before the blinding factor is applied) to fully utilize the space. This is useful, since you need this nonce to actually verify partial signatures in the first place (need to spell this out more in the spec draft).

It seems that `R1` and `R2` are each 33 bytes. Given that `s` already takes up 32 bytes of the originally 64-byte-long-field, I'm not sure there's enough space to squeeze in an extra 66 bytes.

Author: So I need to update this still, my last comment argues that we should just use the full 64 bytes here and send the `r` value along side `s`. 

> It seems that R1 and R2 are each 33 bytes. Given that s already takes up 32 bytes of the originally 64-byte-long-field, I'm not sure there's enough space to squeeze in an extra 66 bytes.

Made an error here translating from brain to writing. What I propose we encode is just `r`, which is the x-coordinate of the final `R` point used for the partial signature generation (the value encoded in the challenge hash). This is a field element, so it's 32 bytes, `s` (bounded by the group order) is also 32 bytes, so everythign fits.

Author: > and therefore the deserialization of the funding_created and funding_signed messages must depend on whether or not they're referring to Taproot channels?

Yes that's one tradeoff with re-using the field vs either: adding another field, or another message all together (would propagate to the ones you mention and the state machine ones as well). On the wire they can both be 64 bytes, but the mapping back to some sort of signature struct/object differs:
  * Schnorr: we have `(r, s)` one is a field element (x coord of the "final" nonce), the other bounded by the group size. 
  * ECDSA: we have `(r, s)` both values being `mod n` (bounded by group size) 

If we go this route, in Go, I'd likely do something like:
```go
type Sig interface {
    Verify(pub *btcec.PublicKey) error
    Encode(w io.Writer) error
    Decode(r io.Reader) error 
}

type CommitSig[S Sig] struct {
     Sig S 
}
```

Then depending on the context, I'd do:
  * regular chans: `CommitSig[*ecdsa.Signature]`
  * taproot chans: `CommitSIg[*schnorr.Signature]` 

With the type bubbling up to other fields where needed.

Reviewer 16: So one major use case of having the aggregate combined final nonce (I opened a [naming issue](https://github.com/jonasnick/bips/issues/69), the ambiguity is killing me) is that its polarity dictates the sign of the secret nonce summands [here](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#signing). With 64 bytes in the `signature` field, only 32 would remain for `r`, so it would have to be an x-only-pubkey, and lose this particular utility of the sanity check.

Further, considering that it would neither be a valid signature, nor a type-conformant partial signature, I do feel like it would be misappropriating the field somewhat. My personal preference, if at the very least for the purpose of clarity and legibility for future protocol readers and developers, would be to introduce TLV extensions for these values, which would also allow the inclusion of the full 66-byte aggregate R-value-serialization.

Reviewer 4: > So I need to update this still, my last comment argues that we should just use the full 64 bytes here and send the `r` value along side `s`.

Why? the final `r` value that could be published to the blockchain (i mean the final combined aggregated nonce or whatever it is called) can be calculated by both parties using the exchanged nonces, one party would never use the nonce he got alongside a `commit_sig` message. i am in favor of just only using 32 (the first 32 or the last 32) bytes of the space available with the partial sig. the other 32 should be stuffed with zeros or at least not untouched data from the ram (like c `malloc` or js `Buffer.allocUnsafe` do)

Author: > Why? the final r value that could be published to the blockchain (i mean the final combined aggregated nonce or whatever it is called) can be calculated by both parties using the exchanged nonces, one party would never use the nonce he got alongside a commit_sig message

The idea is that we might as well just use the space for something useful. This is the value that's needed to combine the partial signatures into an actual signature. Sending it is just a sort of echo that lets both sides verify this value. 

Alternatively, we could send the nonce that's needed to verify the partial signature sent. This would sorta be more useful, since a party can take this value directly w/o needing to maintain some extra state to know which nonce to use.

Author: > would be to introduce TLV extensions for these values, which would also allow the inclusion of the full 66-byte aggregate-nonce-serialization.

Well in order to generate a sig in the first place, both sides already need this value (which I think is the argument being made above). So then we'd be sending extra unnecessary bytes when we can just send 64-bytes minimally with useful information. 

If we are sending a value, then the most useful value would likely be the same nonce one needs to verify the incoming partial signature.

Reviewer 16: Yeah, I was basically saying that I'd prefer the extra bytes to overloading the (de)serialization mechanism depending on the channel type; disregarding encoding, however, how should a node handle an `R` value that deviates from its expectation?

    aggregated key defined above, and the two public nonces exchanged in prior
    messages.

  - MUST generate the attached `signature` using the combined
    `remote_musig2_pubnonce` it generated in the prior step.

> The recipient:

Author: Sort of captured in the current requirement, no? If the aggregated key is different that expected, the sig validation will fail.

> The recipient:

Reviewer 5: The key isn't mentioned at all in the recipient section, so even if "obvious" I think saying which public key you're verifying against is required.

>   - MUST validate the signature as a _partial_ signature based on the

Reviewer 5: partial sigs are 32 bytes, not 64 like full sigs. Are your requiring the public nonces to be re-sent or should we break this into a new message type that's more precisely defined?

Author: See this comment: https://github.com/lightning/bolts/pull/995#issuecomment-1170703114

Main motivation to changing it is sending less bytes I guess? Tho it pales in comparison to how large the onoin messages are.

    `PartialSigVerifyInternal` (using the combined nonce for the remote party's
    commitment) algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

#### `funding_signed` Extensions

The sender MUST set:

  - `signature` to the valid signature, using its `funding_pubkey` for the
    initial commitment transaction, as defined in this documewnt.
    - this signature MUST be a partial signature as defined by `bip-musig2`.

The recipient:

  - MUST validate the signature as a _partial_ signature based on the
    `PartialSigVerifyInternal` algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

#### `funding_locked` Extensions

We add a new TLV field to the `funding_locked` message:

1. type: 36 (`funding_locked`)
2. data:
    * ...
    * [`funding_locked_tlvs`:`tlvs`]

1. `tlv_stream`: `funding_locked_tlvs`
2. types:
    1. type: 0 (`local_musig2_pubnonce`)
    2. data:
        * [`musig2_pubnonce`:`nonces`]
    1. type: 2 (`remote_musig2_pubnonce`)
    2. data:

>         * [`musig2_pubnonce`:`nonces`]

Reviewer 15: These nonces only need to sent the first time `funding_locked` is sent iff a `channel_reestablish` has not already been exchanged. With #910 , the `funding_locked` message would continually replace the nonces, possibly messing up the channel state since the sender of `funding_locked` may not know whether a received `commit_sig` is using the right nonces due to the lack of synchronization. So I think these nonces should actually be exchanged in `funding_created`/`funding_signed`

Author: Good point, I was thinking of the case where a party sends `funding_locked` more than once to update/refresh their scid aliases... 

The goal here was to mirror the existing `funding_locked` behavior and disallow creating a new commitment until this is sent. However this message still has the next revocation key, so you can't make a state w/o that anyway. 

But if we removed the nonces here, then both sides would need to remember the nonce state between restarts. So we might _just_ want to specify that the nonces only need to be in place the very first time chan reest was sent.

Similar to the `next_per_commitment_point`, by sending the `musig2_pubnonce`
value in this message, we ensure that the remote party has our public nonce
which is required to generate a new commitment signature. Similarly, we can't
generate a new signature until we receive their `local_musig2_pubnonce` and
`remote_musig2_pubnonce`.

##### Requirements

The sender MUST:

 - MUST set `local_musig2_pubnonce` and `remote_musig2_pubnonce` to a _fresh_
   set of public nonces as specified by `bip-musig2`

The receiver MUST:

  - MUST combine their local `pubnonce` and the received `pubnonce` via the
    `NonceAgg` algorithm specified in the `musig2` BIP to received _two_
    combined nonces. This combined nonce will be used to generate the next
    commitment signature via the `Sign` algorithm specified in the `musig2`
    BIP.

### Cooperative Closure

#### `shutdown` Extensions

A new TLV field is added to the `shutdown` message:

1. type: 38 (`shutdown`)
2. data:
    * ...
    * [`shutdown_tlvs`:`tlvs`]

1. `tlv_stream`: `shutdown_tlvs`
2. types:

>     1. type: 2 (`musig2_pubnonce`)

Author: Will make this change once https://github.com/Roasbeef/lightning-rfc/pull/2 is merged, or we just do it there perhaps.

    2: data:
        * [`66*byte`:`nonces`]

Before a signature can be generated for a co-op close transaction, but sides
must exchange a fresh pair of `musig2_pubnonce` values. We package this with
the `shutdown` message so that both sides can send a `closing_signed` message
as soon as a `shutdown` message is sent and received.

Only a single nonce is needed as there's only a single message to sign: the
shared co-op close transaction.

##### Requirements

A sending node:

 - MUST set the `musig2_pubnonce` to a valid `musig2` public nonce.

A receiving node:

 - MUST verify that the `musig2_pubnonce` value is a valid `musig2` public nonce.

#### `closing_signed` Extensions

Similar to the `shutdown` message, we add a new TLV to the `closing_signed`
message's existing `tlv_stream`:

1. `tlv_stream`: `closing_signed_tlvs`
2. types:

>     1. type: 2 (`musig2_pubnonce`)

Reviewer 15: These will need 2 nonces (our local, our remote) - it's not possible to totally eliminate signatures until the last step unless we add an extra redundant `closing_signed` (which lnd does, but isn't in the spec)

>     1. type: 2 (`musig2_pubnonce`)

Reviewer 15: I made a pr to make `fee_range` turn-based for taproot here: https://github.com/Roasbeef/lightning-rfc/pull/2

    2: data:
        * [`66*byte`:`nonces`]

Both sides **MUST** provide this new TLV field.

As a fresh `musig2_pubnonce` is required to sign another fee proposal, with
each new signature, both sides send a new nonce that'll be used for the _next_
signature.

### Channel Operation

#### `commit_sig` Extensions

The main commitment `signature` attached in the message is now a `bip-musig2`
_partial_ schnorr signature. Using the `secnonce` generated upon connection
re-establishment or funding confirmation, the partial signature is generated
with the `Sign` algorithm specified in the `musig2` BIP.

##### Requirements

The sender MUST:

 - Generate the commitment signature attached using the `Sign` algorithm as
   specified in the `musig2` BIP, the aggregated public key generating during
   funding, and the local `secnonce` along with the combined aggregated public
   nonce.

>  - Generate each HTLC signature according to BIP 342, as a _normal_ BIP 340

Reviewer 5: well, a BIP340 sig is always a "normal" sig, so drop the word I guess

>  - Generate each HTLC signature according to BIP 342, as a _normal_ BIP 340

Author: Removed normal, added "non partial" in parens.

the spending conditions do not require a script path, the output key should
commit to an unspendable script path instead of having no script path. This can
be achieved by computing the output key point as `Q = P +
int(hashTapTweak(bytes(P)))G`". We refer to BIP 86 for the specifics of this
derivation.

By using the musig2 sorting routine, we inherit the lexicographical key
ordering from the base segwit channels. Throughout the lifetime of a channel
any keys aggregated via musig2 also are assumed to use the `KeySort` routine

### Commitment Transactions

#### To Local Outputs

For the simple taproot commitment type, we use the same flow of revocation or
delay, but instead left the revocation case into the taproot key-spend case.

The new output has the following form:

  * `OP_1 to_local_output_key`
  * where:
    * `to_local_output_key = <revocationpubkey> + tagged_hash("TapTweak", <revocationpubkey> || to_delay_script_root`
    * `to_delay_script_root = tapscript_root([to_delay_script])`
    * `to_delay_script` is the delay script:
        ```
        to_self_delay
        OP_CHECKSEQUENCEVERIFY
        OP_DROP
        <local_delayedpubkey>

>         OP_CHECKSIGVERIFY

Reviewer 4: fixed in https://github.com/Roasbeef/lightning-rfc/pull/1

and the `remotepubkey` rather than the multi-sig keys as that's no longer
revealed due to musig2.

An anchor output has the following form:

  * `OP_1 anchor_output_key`
  * where:
    * `anchor_internal_key = remotepubkey/local_delayedpubkey`
    * `anchor_output_key= anchor_internal_key + tagged_hash("TapTweak", anchor_internal_key || anchor_script_root)`
    * `anchor_script = tapscript_root([anchor_script])`
    * `anchor_script`:
        ```
        OP_16 OP_CHECKSEQUENCEVERIFY
        ```

### HTLC Scripts & Transactions

#### Offered HTLCs

We retain the same second-level HTLC mappings as base channels. The main
modifications are using the revocation public key as the internal key, and
splitting the timeout and success paths into individual script leaves.

An offered HTLC has the following form:

  * `OP_1 offered_htlc_key`
  * where:
    * `offered_htlc_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
    * `htlc_script_root = tapscript_root([htlc_timeout, htlc_success])`
    * `htlc_timeout`:
        ```
        <local_htlcpubkey> OP_CHECKSIGADD <remote_htlcpubkey> OP_CHECKSIGADD 2 OP_EQUALVERIFY

>         1 OP_CHECKSEQUENCEVERIFY

Reviewer 4: `OP_CHECKSIGADD` is added as a replacement for t of n multisignatures where t is not n, simple n of n multisignatures can use `<P1> OP_CHECKSIGVERIFY <P2> OP_CHECKSIGVERIFY ... <Pn> OP_CHECKSIG` (checksig only for the last one)

btw this script is anyone can spend with `<> <> <2>`

https://github.com/Roasbeef/lightning-rfc/pull/1

Reviewer 15: Oh shit, nice catch - these should be `OP_CHECKSIGVERIFY` then @Roasbeef

        ```
    * `htlc_success`:
        ```
        OP_SIZE 32 OP_EQUALVERIFY OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY
        <remote_htlcpubkey>
        OP_CHECKSIGVERIFY
        1 OP_CHECKSEQUENCEVERIFY
        ```

In order to spend a offered HTLC, via either script path, an `inclusion_proof`
must be specified along with the control block. This `inclusion_proof` is
simply the `tap_leaf` hash of the path _not_ taken.

#### Accepted HTLCs

Accepted HTLCs inherit a similar format:

  * `OP_1 accepted_htlc_key`
  * where:
    * `accepted_htlc_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
    * `htlc_script_root = tapscript_root([htlc_timeout, htlc_success])`
    * `htlc_timeout`:
        ```
        <remote_htlcpubkey>
        OP_CHECKSIGVERIFY
        1 OP_CHECKSEQUENCEVERIFY OP_DROP
        <cltv_expiry>

>         OP_CHECKLOCKTIMEVERIFY

Reviewer 4: fixed in https://github.com/Roasbeef/lightning-rfc/pull/1

>         OP_CHECKLOCKTIMEVERIFY

Author: I think the latest variant of the script allows such a value?

        ```
    * `htlc_success`:
        ```
        OP_SIZE 32 OP_EQUALVERIFY OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY

>         <local_htlcpubkey> OP_CHECKSIGADD <remote_htlcpubkey> OP_CHECKSIGADD 2 OP_EQUALVERIFY

Reviewer 15: These need to be changed similarly to enforce the spending tx is the 2nd-level tx

Reviewer 4: fixed in https://github.com/Roasbeef/lightning-rfc/pull/1

multi-spend going to a CTV delayed output. Like the normal HTLC transactions,
we also place the revocation key as the internal key, allowing easy breach
sweeps with each party only retaining the specific script root.

An HTLC-Success or HTLC-Timeout is a one-in-one-out transaction signed using
the `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` flag. These transactions always have
_zero_ fees attached, forcing them to be aggregated with each other and a
change input.

##### HTLC-Success Transactions

A HTLC-Success transaction has the following structure:

  * version: 2
  * locktime: 0
  * input:
    * txid: commitment_tx
    * vout: htlc_index
    * witness: `control_block || htlc_success_script || <localsig> <remotesig> <preimage>`
  * output:
    * value: htlc_value
    * script:
      * OP_1 htlc_success_key
      * where:
        * `htlc_success_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
        * `htlc_script_root = tapscript_root([htlc_success])`
        * `htlc_success`:
        ```
        <local_delayedpubkey>
        OP_CHECKSIGVERIFY
        <to_self_delay>

>         OP_CHECKSEQUENCEVERIFY

Reviewer 4: fixed in https://github.com/Roasbeef/lightning-rfc/pull/1

>         OP_CHECKSEQUENCEVERIFY

Author: I think the latest script permits this since a 1 will be left on the stack?

        ```

##### HTLC-Timeout Transactions

A HTLC-Timeout transaction has the following structure:

  * version: 2
  * locktime: cltv_expiry
  * input:
    * txid: commitment_tx
    * vout: htlc_index
    * witness: `control_block || htlc_timeout_script || <localsig> <remotesig>`
  * output:
    * value: htlc_value
    * script:
      * OP_1 htlc_timeout_key
      * where:
        * `htlc_timeout_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
        * `htlc_script_root = tapscript_root([htlc_timeout])`
        * `htlc_timeout`:
        ```
        <local_delayedpubkey>
        OP_CHECKSIGVERIFY
        <to_self_delay>

>         OP_CHECKSEQUENCEVERIFY

Reviewer 4: fixed in https://github.com/Roasbeef/lightning-rfc/pull/1

will be absent.

The final witness to spend a script path output is:
```
<witness1> ... <witnessN> <leaf_script> <control_block>
```

### MuSig2

Musig2 is a multi-signature scheme based on schnorr signatures. It allows N
parties to construct an aggregated public key, and then generate a single
multi-signature that can only be generated with all the parties of the
aggregated key. In practice we use this to allow the funding output of
channel to look just like any other P2TR output.

A musig2 signing flow has 4 stages:

  1. First each party generates a public nonce, and exchanges it with every
    other party.

  2. Next public keys are exchanged by each party, and combined into a single
    aggregated public key.

  3. Next, each party generates a _partial_ signature and exchanges it with
    every other party.

  4. Finally, once all partial signatures are known, they can be combined into a
    valid schnorr signture, which can be veirfied using BIP 340.

Steps 1&2 may be performed out of order, or concurrently. In our case only two
parties exist, so as soon as one party knows both partial signatures, they can
be combined into a final signature.

#### Key Aggregation

> We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)

Author: > Might affect some part of this proposal but haven't checked it in full yet.

It'll affect the ultimate test vectors, and interop will require both sides being on the same musig2 version (ofc). `btcd` has implemented up to version 0.4.0 so far, and this is also what's packaged in version 0.15.0 of `lnd` w/ an experimental RPC API. 

I think so far only the Python ref implementation is on version 0.8.1, afact secp-zkp is on 0.3.0 (or so).

> We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)

Author: I think by the time this is ready for prime time (cross impl interop, etc) we'll ideally be at musig 1.0. Then we'll just pin that and we can safely make test vectors to fill in everything.

> We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)

Reviewer 8: I believe we can update all the `bip-musig2` references to `BIP 327` since we're [now at 1.0.0](https://github.com/bitcoin/bips/blob/b4ea0a93aaa469057cd88695c8d56ef92a6734b3/bip-0327.mediawiki?plain=1#L788).

for key aggregation. In order to avoid sending extra ordering information, we
always assume that both keys are sorted first with the `KeySort` algorithm
before aggregating (`KeyAgg(KeySort(p1, p2))`)>

#### Nonce Generation

A `musig2` public nonce is a 66-byte value which is the concentration of two EC
points encoded using the compressed format:
```
nonce_1 || nonce_2
```

Each public nonce has a corresponding _secret_ nonce which is only known to the
signer.

For nonce generation, we refer to the `NonceGen` algorithm defined in the
`musig2` BIP. We only specify the set of required arguments, though a signer
can also opt to specify the additional argument in order to strengthen their
nonce.

#### Nonce Handling

Throughput this document we always assume that _fresh_ nonces are generated for
each signing attempt, and nonces are **never** written to disk. As we'll see
later, this has implications with respect to the retransmission protocol.

> Note that nonces can safely be pre-generated, and even batched/pipelined ahead

Reviewer 17: It might be worth noting that to "safely" pre-generate the nonces, some care needs to be taken to ensure a pre-generated nonce is never reused. As noted in the musig [doc](https://github.com/ElementsProject/secp256k1-zkp/blob/master/src/modules/musig/musig.md): Pre-generating nonces "disables some of the defense-in-depth measures that may protect against API misuse in some cases."

Reviewer 5: a straight link to the BIP is probably a good idea here to know exactly what should be raising red flags when implementing. e.g. I find it highly sketchy if someone thinks they need to serialize the secret nonce.

Author: Agreed, it ultimately depends on what the `musig` API of the implementations looks like. By "pre-generate" nonces, I mean generating your nonce before you know the public key of the remote party. 

Will update this with some links to the BIP, and maybe also will start to directly reference some of the algos there.

Author: Tacked on a section here describing the purely random nonce generation, and also a way to used the existing commitment height/number and the shachain scheme to generate a unique nonce for each channel state.

Anchor outputs use the respective local and remote keys from the commitment
transaction as the top level output, with the committed script being the `CSV
16` delay.

All HTLC scripts use the revocation key as the top-level internal key. This
allows the revocation to be executed without additional on-chain bytes, and
also reduces the amount of state nodes need to keep in order to properly handle
channel breaches.

## Specification

### Feature Bits

Inheriting the structure put forth in BOLT 9, we define a new feature bit to be
placed in the `init` message, `node_announcement` and also
`channel_announcement`. We opt to also place the bit in the
`channel_announcement` so taproot channels can be identified on the routing
level, which will eventually matter once PTLCs are fully rolled out.

| Bits  | Name                             | Description                                               | Context  | Dependencies      | Link                                  |
|-------|----------------------------------|-----------------------------------------------------------|----------|-------------------|---------------------------------------|

> | 50/51 | `option_simple_taproot`| Node supports simple taproot chanenls | IC | `option_channel_type`+`option_anchors` | TODO(roasbeef): link |

Reviewer 16: does 50/51 not conflict with option_zeroconf?

Author: Now 80/81 and also 180/181 (staging bit).

    received from the remote party with the `NonceAgg` algorithm to derive the
    combined nonce

  - the `signature` field MUST be a `musig2` partial signature using the
    aggregated key defined above, and the two public nonces exchanged in prior
    messages.

  - MUST generate the attached `signature` using the combined
    `remote_musig2_pubnonce` it generated in the prior step.

The recipient:
  - MUST validate the signature as a _partial_ signature based on the
    `PartialSigVerifyInternal` (using the combined nonce for the remote party's
    commitment) algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

#### `funding_signed` Extensions

The sender MUST set:

  - `signature` to the valid signature, using its `funding_pubkey` for the
    initial commitment transaction, as defined in this documewnt.
    - this signature MUST be a partial signature as defined by `bip-musig2`.

The recipient:

  - MUST validate the signature as a _partial_ signature based on the
    `PartialSigVerifyInternal` algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

> #### `funding_locked` Extensions

Reviewer 16: I believe this may need to be renamed to `channel_ready`

# Extension BOLT XX: Simple Taproot Channels

Authors:
  * Olaoluwa Osuntokun <roasbeef@lightning.engineering>
  * Eugene Siegel <eugene@lightning.engineering>

Created: 2022-04-20

# Table of Contents

- [Introduction](#introduction)
  * [Abstract](#abstract)
  * [Motivation](#motivation)
  * [Preliminaries](#preliminaries)
    + [Pay-To-Taproot-Outputs](#pay-to-taproot-outputs)
      - [Tapscript Tree Semantics](#tapscript-tree-semantics)
      - [BIP 86](#bip-86)
      - [Taproot Key Path Spends](#taproot-key-path-spends)
      - [Taproot Script Path Spends](#taproot-script-path-spends)
    + [MuSig2](#musig2)
      - [Key Aggregation](#key-aggregation)
      - [Nonce Generation](#nonce-generation)
      - [Nonce Handling](#nonce-handling)
      - [Signing](#signing)
    + [Nothing Up My Sleeves Points](#nothing-up-my-sleeves-points)
  * [Design Overview](#design-overview)
  * [Specification](#specification)
    + [Feature Bits](#feature-bits)
    + [Channel Funding](#channel-funding)

>       - [`accept_channel` Extensions](#-accept-channel--extensions)

Reviewer 16: this should probably be open channel extensions. Also, the link should be `#open_channel-extensions`

>       - [`accept_channel` Extensions](#-accept-channel--extensions-1)

Reviewer 16: similarly, here, Github wants the link to be `#accept_channel-extensions`

      - [`funding_created` Extensions](#-funding-created--extensions)
      - [`funding_signed` Extensions](#-funding-signed--extensions)

>       - [`funding_locked` Extensions](#-funding-locked--extensions)

Reviewer 16: I believe it's `channel_ready` since 0conf

to the remote party.

A peer's verification nonce is used to verify incoming commitments sent by the
remote party. Once a peer verifies and revoke a new commitment, another
verification nonce is sent to the remote party.

Both sides will maintain a `musig2` sessions for their local commitment and the
commitment of the remote party. A peer's local commitment uses their
verification nonce, and the signing nonce of the remote party. The remote
commitment of the channel peer uses the peer's signing nonce, and their
verification nonce.

In order to sign and broadcast a new commitment to force close a channel, a
peer uses their _verification_ nonce to generate their final signature.

#### Signing

Once public nonces have been exchanged, the `NonceAgg` algorithm is used to
combine the public nonces into the combined nonce which will be a part of the
final combined signature.

The `Sign` is used to generate a valid partial signature, with the
`PartialSigVerify` algorithm used to verify each signature.

Once all partial signatures are known, the `PartialSigAgg` algorithm is used to
combine the partial signature into a final valid BIP 340 schnorr signature.

> ##### Partial Signature Encoding

Reviewer 18: ```suggestion
##### [Partial Signature Encoding](#partial-signature-encoding)
```

A `musig2` partial signature is simply the encoding of the `s` value which is
32 byte element modulo the order of the group. Throughout this specification,
in order to be able to re-use the existing 64-byte signature field, partial
signatures are also encoded with 64 bytes by also encoding the x-coordinate of
the _final_ nonce (`R1+b*R2`).

## Design Overview

With the preliminaries out of the way, we provide a brief overview of the
Simple Taproot Channel design.

The multi-sig output becomes a single P2TR key, with `musig2` key aggregation
used to combine two keys into one. During the funding flow, public nonces are
exchanged with the first two messages.

For commitment transactions, we inherit the anchor outputs semantics, meaning
there're two outputs used for CPFP, with all scripts inheriting a `CSV 1`
restriction.

> The local output of the commitment transaction uses the revocation key as the
> taproot internal key, and commits to a single script path of the delay script.

Reviewer 18: ```suggestion
The local output, of the commitment transaction, uses the revocation key as the
taproot internal key and commits to a single script path of the delay script.
```

> The remote output of the commitment transaction uses the `combined_funding_key`
> as the top-level internal key, and then commits to a normal remote script.

Reviewer 18: ```suggestion
The remote output, of the commitment transaction, uses the `combined_funding_key`
as the top-level internal key and then commits to a normal remote script.
```

Anchor outputs use the respective local and remote keys from the commitment
transaction as the top level output, with the committed script being the `CSV
16` delay.

> All HTLC scripts use the revocation key as the top-level internal key. This
> allows the revocation to be executed without additional on-chain bytes, and
> also reduces the amount of state nodes need to keep in order to properly handle
> channel breaches.

Reviewer 18: ```suggestion
All HTLC scripts use the revocation key as the top-level internal key. This
allows the revocation transaction to be executed without additional on-chain
bytes, and also reduces the amount of state nodes need to keep in order to
properly handle channel breaches.
```

local commitment and the remote commitment. If only one nonce was used, when a
party went to sign and broadcast their local commitment, they would re-use a
nonce, as the nonce was already used to sign the remote party's commitment.

#### `accept_channel` Extensions

1. `tlv_stream`: `accept_channel_tlvs`
2. types:
    1. type: 2 (`verification_musig2_pubnonce`)
    2: data:
        * [`66*byte`:`nonces`]
    1. type: 4 (`signing_musig2_pubnonce`)
    2: data:
        * [`66*byte`:`nonces`]

##### Requirements

The sender:

 - MUST set `verification_musig2_pubnonce` and `signing_musig2_pubnonce` to the
   `musig2` public nonce as specified by the `NonceGen` algorithm of
   `bip-musig2`

 - Aggregate the received local+remote public nonces (into two combined nonces)
   using the `NonceAgg` algorithm of `bip-musig2`.

The receiver:

 - MUST reject the channel if `signing_musig2_pubnonce` or
   `verification_musig2_pubnonce` are not present.

#### `funding_created` Extensions

The sender:

  - MUST derive the aggregated `musig2` public key using the exchanged `funding_pubkey`s
    - Before aggregating the two funding public keys, the keys MUST be sorted

>       using the `KeySort` routine of `bip-musig2`.

Author: Now refers to BIP 327.

  - MUST use the exchanged `signing_musig2_pubnonce` and
    `verification_musig2_pubnonce` received from the remote party with the
    `NonceAgg` algorithm to derive the combined nonce. 
    - Local commitment combined nonce: `NonceAgg(verificationNonce, signingNonce)`
        - Where `verificationNonce` is the local party's and `signingNonce` is
          the remote party's.
    - Remote Commitment combined nonce: `NonceAgg(signingNonce, verificationNonce)`
        - Where `signingNonce` is the local party's and `verificationNonce` is
          the remote party's.
  - the `signature` field MUST be a `musig2` partial signature using the

>     aggregated key defined above (serialized according to the Partial Signature
>     Encoding section), and the two public nonces exchanged in prior messages.

Reviewer 18: ```suggestion
    aggregated key defined above (serialized according to the [Partial Signature
    Encoding section](#partial-signature-encoding)), and the two public nonces exchanged in prior messages.
```

  - MUST generate the attached `signature` using the combined
    Remote commitment combined nonce it generated in the prior step.

The recipient:

>   - MUST validate that the 

Reviewer 18: incomplete

  - MUST validate the signature as a _partial_ signature based on the
    `PartialSigVerifyInternal` (using the verification nonce of the local
    party) algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

#### `funding_signed` Extensions

The sender MUST set:

  - `signature` to the valid signature, using its `funding_pubkey` for the
    initial commitment transaction, as defined in this documnt.
    - this signature MUST be a partial signature as defined by `bip-musig2`.

The recipient:

  - MUST validate the signature as a _partial_ signature based on the
    `PartialSigVerifyInternal` algorithm of `bip-musig2`:

    - if the partial signature is invalid, MUST fail the channel

#### `funding_locked` Extensions

We add a new TLV field to the `funding_locked` message:

1. type: 36 (`funding_locked`)
2. data:
    * ...
    * [`funding_locked_tlvs`:`tlvs`]

1. `tlv_stream`: `funding_locked_tlvs`
2. types:

>     1. type: 0 (`verification_musig2_pubnonce`)

Reviewer 16: should the types here remain 0 and 2, where in all other extensions, they're 2 and 4?

Author: As is the nonce values are static, so `4` is always used for the nonce type. `2` is always the type for partial sig with nonce, and `6` is always used for the partial signature type, and `8` is used for the shutdown nonces.

    2. data:
        * [`musig2_pubnonce`:`nonces`]
    1. type: 2 (`signing_musig2_pubnonce`)
    2. data:
        * [`musig2_pubnonce`:`nonces`]

Similar to the `next_per_commitment_point`, by sending the `musig2_pubnonce`
value in this message, we ensure that the remote party has our public nonce
which is required to generate a new commitment signature. Similarly, we can't
generate a new signature until we receive their `verification_musig2_pubnonce`
and `signing_musig2_pubnonce`.

##### Requirements

The sender MUST:

 - MUST set `verification_musig2_pubnonce` and `signing_musig2_pubnonce` to a
   _fresh_ set of public nonces as specified by `bip-musig2`

The receiver MUST:

  - MUST combine their local `pubnonce` and the received `pubnonce` via the
    `NonceAgg` algorithm specified in the `musig2` BIP to received _two_
    combined nonces. This combined nonce will be used to generate the next
    commitment signature via the `Sign` algorithm specified in the `musig2`
    BIP.

### Cooperative Closure

> #### `shutdown` Extensions

Reviewer 7: if shutdown is being kinda re-done, it would be nice if we could fit something in here to allow for adding extra inputs & outputs, ie PSBT funding flow but for channel closes

Reviewer 10: I agree, in the last spec meeting we discussed using the `interactive-tx` protocol (used in dual funding) to collaboratively build two mutual close transactions in parallel (one for each peer, with its own preferences about feerate and other inputs / outputs). This sounds like the right way to do closing in the future, but it does require implementations to already support `interactive-tx`.

So we may take a first intermediate step with a simpler, smaller change to the closing protocol before we eventually move to `interactive-tx`.

Author: Latest version moves the co-op close transaction to always signal RBF. This opens the door to the new simple coop close proposal, and future versions that use `interactive-tx` to make diff RBF'd version to add additional inputs/outputs.

along side `<revocationpubkey>` to derive the private key needed to sweep the
top-level key spend path. A valid witness is then just:
```
<revocation_sig>
```

In the case of a routine force close, the script path must be revealed so the
broadcaster can sweep their funds after a delay. The control block to spend is
only `33` bytes, as it just includes the internal key (along with the y-parity
bit and leaf version):
```
delay_control_block = (output_key_y_parity | 0xc0) || revocationpubkey
```

A valid witness is then:
```
<local_delayedsig> <to_delay_script> <delay_control_block>
```

As with base channels, the `nSequence` field must be set to `to_self_delay`.

#### To Remote Outputs

As we inherit the anchor output semantics we want to ensure that the remote
party can unilaterally sweep their funds after the 1 block CSV delay. In order
to achieve this property, we'll re-use the `_funding_key` here: its in
the best interest of the other party to enforce these semantics (mempool
pinning mitigation), so the remote party will be forced to always take the
script reveeal path.

The to remote output has the following form:

>   * `OP_1 to_remote_output_key`

Reviewer 16: What exactly do you mean by this? A Taproot output, as far as I understand, can only have the form `0x51 <32-byte-pubkey>`. This looks like there's also an `OP_1` prefix for the `OP_CHECKSEQUENCEVERIFY` command, but it seems to exist outside the tap tree structure? Can you maybe give an example with sum pubkeys of what the Taproot Merkle tree and output script would look like?

Reviewer 4: I dont see what is wrong with this. OP_1 == 0x51 and the 32 byte pubkey in this case is the to remote output key

Reviewer 16: My confusion partially stems from the `to_remote_script` semantics. In BOLT 3, the script is `<remotepubkey> OP_CHECKSIGVERIFY 1 OP_CHECKSEQUENCEVERIFY`. Here, however, the input to `OP_CSV` seems to be dependent on the validity of the signature against `remotepubkey`. So if an invalid signature is presented, the script stack would end up as `0 OP_CSV`?

Reviewer 4: The script can only succeed with a valid signature, and then the sequence is checked. The OP_1 from the taproot _output_ has nothing to do with the 1 block relative timelock

Reviewer 16: Right, disregarding the OP_1 (which is indeed simply the witness version, sorry for the distraction), my concern was that `O OP_CSV` wouldn't fail, but given that even though the `OP_CSV` part doesn't fail per se, it leaves a falsey stack element at the top, so all does check out in the end. I'll close this entire thread.

  * where:

>     * `to_remote_output_key = combined_funding_key + tagged_hash("TapTweak", combined_funding_key || to_remote_script_root)`

Reviewer 16: Additionally, my understanding of Taproot tweak calculations is that the preimage first contains the Merkle root, and then the internal key, so it would actually be
`tweak = tagged_hash("TapTweak", to_remote_script_root || combined_funding_key)`

  * where:
    * `accepted_htlc_key = revocation_pubkey + tagged_hash("TapTweak", revocation_pubkey || htlc_script_root)`
    * `htlc_script_root = tapscript_root([htlc_timeout, htlc_success])`
    * `htlc_timeout`:
        ```
        <remote_htlcpubkey> OP_CHECKSIG
        OP_CHECKSEQUENCEVERIFY
        <cltv_expiry> OP_CHECKLOCKTIMEVERIFY OP_DROP
        ```
    * `htlc_success`:
        ```
        OP_SIZE 32 OP_EQUALVERIFY OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY
        <local_htlcpubkey> OP_CHECKSIGVERIFY
        <remote_htlcpubkey> OP_CHECKSIG
        ```

Note that there is no `OP_CHECKSEQUENCEVERIFY` in the offered HTLC's timeout path
and in the accepted HTLC's success path. This is not needed here as these paths
require a remote signature that commits to the sequence, which needs to be set at 1.

In order to spend an accepted HTLC, via either script path, an
`inclusion_proof` must be specified along with the control block. This
`inclusion_proof` is simply the `tap_leaf` hash of the path _not_ taken.

TODO(roasbeef): specify full witnesses?

#### HTLC Second Level Transactions

For the second level transactions, we retain the existing structure: a 2-of-2

> multi-spend going to a CTV delayed output. Like the normal HTLC transactions,

Reviewer 2: CLTV

> multi-spend going to a CTV delayed output. Like the normal HTLC transactions,

Author: Must've been living in the future 😎

Once all partial signatures are known, the `PartialSigAgg` algorithm is used to
combine the partial signature into a final valid BIP 340 schnorr signature.

##### Partial Signature Encoding

A `musig2` partial signature is simply the encoding of the `s` value which is
32 byte element modulo the order of the group. Throughout this specification,
in order to be able to re-use the existing 64-byte signature field, partial
signatures are also encoded with 64 bytes by also encoding the x-coordinate of

> the _final_ nonce (`R1+b*R2`).

Reviewer 16: (the spec defines this as the "effective" nonce, per https://github.com/jonasnick/bips/issues/69#issuecomment-1244322742) :)

Reviewer 16: I stand corrected, actually, so nvm lol
